{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dc6a73f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Setting >>>>>>>\n",
    "\n",
    "def warn(*args, **kwargs):\n",
    "    pass\n",
    "import warnings\n",
    "warnings.warn = warn\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import gseapy\n",
    "import scvi\n",
    "import scipy.sparse as sp\n",
    "import math\n",
    "import anndata as ad\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "plt.rcParams[\"pdf.fonttype\"] = 42          \n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "\n",
    "import matplotlib.patheffects as pe\n",
    "\n",
    "sc.settings.verbosity = 3 \n",
    "\n",
    "import os\n",
    "print(os.getcwd())\n",
    "cwd = f\"{os.getcwd()}\"\n",
    "\n",
    "sc.settings.set_figure_params(dpi=200)\n",
    "\n",
    "print(matplotlib.rcParams[\"font.family\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b82e604",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< selected gene count >>>>>>>\n",
    "\n",
    "h5ad_path = f\"{cwd}/Neuron2022_Human_TG.h5ad\"\n",
    "out_xlsx  = f\"{cwd}/selected_gene_counts.xlsx\"\n",
    "\n",
    "genes_want = [\"TRAC\", \"TRBC1\", \"TRBC2\", \"CD4\", \"CD8A\", \"CD8B\"]  \n",
    "\n",
    "use_raw = True          \n",
    "layer_name = None       \n",
    "\n",
    "adata = sc.read_h5ad(h5ad_path)\n",
    "\n",
    "if use_raw and adata.raw is not None:\n",
    "    M = adata.raw.X\n",
    "    var_index = adata.raw.var_names\n",
    "else:\n",
    "    if layer_name is not None:\n",
    "        if layer_name not in adata.layers:\n",
    "            raise ValueError(f\"layer '{layer_name}' did not found. Available: {list(adata.layers.keys())}\")\n",
    "        M = adata.layers[layer_name]\n",
    "    else:\n",
    "        M = adata.X\n",
    "    var_index = adata.var_names\n",
    "\n",
    "var_index = pd.Index(var_index)\n",
    "found_mask = var_index.isin(genes_want)\n",
    "genes_found = var_index[found_mask].tolist()\n",
    "genes_missing = sorted(set(genes_want) - set(genes_found))\n",
    "\n",
    "if len(genes_found) == 0:\n",
    "    raise ValueError(\"The specified gene was not found. Please check the spelling in var_names.\")\n",
    "\n",
    "gene_idx = var_index.get_indexer(genes_found)\n",
    "\n",
    "if sp.issparse(M):\n",
    "    sub = M[:, gene_idx].toarray()\n",
    "else:\n",
    "    sub = M[:, gene_idx]\n",
    "\n",
    "df = pd.DataFrame(sub, index=adata.obs_names, columns=genes_found)\n",
    "\n",
    "with pd.ExcelWriter(out_xlsx, engine=\"xlsxwriter\") as writer:\n",
    "    df.to_excel(writer, sheet_name=\"counts_cells_x_genes\", index=True)\n",
    "\n",
    "    pd.DataFrame({\"missing_genes\": genes_missing}).to_excel(\n",
    "        writer, sheet_name=\"missing_genes\", index=False\n",
    "    )\n",
    "\n",
    "print(\"done:\", out_xlsx)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f31e102",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = adata.var_names.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5692919",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Cell count >>>>>>>\n",
    "\n",
    "adata = sc.read_h5ad(f\"{cwd}/Neuron2022_Human_TG.h5ad\")\n",
    "\n",
    "def get_expr_or_zero(adata, gene):\n",
    "    if gene in adata.var_names:\n",
    "        x = adata[:, gene].X\n",
    "        expr = x.toarray().flatten() if not isinstance(x, np.ndarray) else x.flatten()\n",
    "        return expr > 0\n",
    "    else:\n",
    "        print(f\"[skip] {gene} is not exist\")\n",
    "        return np.zeros(adata.n_obs, dtype=bool)\n",
    "    \n",
    "def get_multi_gene_expr(adata, genes):\n",
    "    return np.logical_or.reduce([get_expr_or_zero(adata, g) for g in genes])\n",
    "\n",
    "exp_TRA = get_multi_gene_expr(adata, [\"TRAC\"])\n",
    "exp_TRB = get_multi_gene_expr(adata, [\"TRBC1\", \"TRBC2\"])\n",
    "exp_CD4 = get_multi_gene_expr(adata, [\"CD4\"])\n",
    "exp_CD8 = get_multi_gene_expr(adata, [\"CD8A\", \"CD8B\", \"CD8B2\"])\n",
    "exp_CD3 = get_multi_gene_expr(adata, [\"CD3D\", \"CD3E\", \"CD3G\"])\n",
    "\n",
    "exp_CSF1R = get_multi_gene_expr(adata, [\"CSF1R\"])\n",
    "exp_ZBTB46 = get_multi_gene_expr(adata, [\"ZBTB46\"])\n",
    "exp_HLA_DRA = get_multi_gene_expr(adata, [\"HLA-DRA\"])\n",
    "exp_H2AB1 = get_multi_gene_expr(adata, [\"H2AB1\"])\n",
    "exp_LYZ = get_multi_gene_expr(adata, [\"LYZ\"])\n",
    "exp_MS4A6A = get_multi_gene_expr(adata, [\"MS4A6A\"]) \n",
    "exp_AIF1 = get_multi_gene_expr(adata, [\"AIF1\"])\n",
    "exp_MERTK = get_multi_gene_expr(adata, [\"MERTK\"])\n",
    "exp_TMEM119 = get_multi_gene_expr(adata, [\"TMEM119\"])\n",
    "exp_SOX9 = get_multi_gene_expr(adata, [\"SOX9\"])\n",
    "\n",
    "exp_CEACAM8 = get_multi_gene_expr(adata, [\n",
    "    \"CEACAM8\", \"LY6G6C\", \"LY6G6D\", \"LY6G6E\", \"LY6G6F\", \n",
    "])\n",
    "\n",
    "exp_TRGV = get_multi_gene_expr(adata, [\n",
    "    \"TRGV1\", \"TRGV2\", \"TRGV3\", \"TRGV4\", \"TRGV5\", \"TRGV8\", \"TRGV9\", \"TRGV10\", \"TRGV11\"\n",
    "])\n",
    "\n",
    "exp_TRDV = get_multi_gene_expr(adata, [\n",
    "    \"TRDV1\", \"TRDV2\", \"TRDV3\"\n",
    "])\n",
    "\n",
    "exp_IFNG = get_multi_gene_expr(adata, [\"IFNG\"])\n",
    "exp_PTPRC = get_multi_gene_expr(adata, [\"PTPRC\"])\n",
    "\n",
    "\n",
    "\n",
    "adata.obs[\"CD4+\"] = exp_CD4 \n",
    "adata.obs[\"CD8+\"] = exp_CD8 \n",
    "adata.obs[\"CD4+_CD8-\"] = exp_CD4 & ~exp_CD8\n",
    "adata.obs[\"CD4-_CD8+\"] = exp_CD8 & ~exp_CD4\n",
    "adata.obs[\"CD4+_CD8+\"] = exp_CD4 & exp_CD8\n",
    "adata.obs[\"CD4-_CD8-\"] = ~exp_CD4 & ~exp_CD8\n",
    "\n",
    "adata.obs[\"TRA+\"] = exp_TRA\n",
    "adata.obs[\"TRB+\"] = exp_TRB\n",
    "adata.obs[\"TRA+_TRB+\"] = exp_TRA & exp_TRB\n",
    "adata.obs[\"(TRA+_or_TRB+)\"] = exp_TRA | exp_TRB\n",
    "\n",
    "adata.obs[\"TRA+_CD4+\"] = exp_TRA & exp_CD4\n",
    "adata.obs[\"TRA+_CD8+\"] = exp_TRA & exp_CD8\n",
    "adata.obs[\"TRB+_CD4+\"] = exp_TRB & exp_CD4\n",
    "adata.obs[\"TRB+_CD8+\"] = exp_TRB & exp_CD8\n",
    "\n",
    "adata.obs[\"TRA+_TRB+_CD4+_CD8-\"] = exp_TRA & exp_TRB & exp_CD4 & ~exp_CD8\n",
    "adata.obs[\"TRA+_TRB+_CD4-_CD8+\"] = exp_TRA & exp_TRB & exp_CD8 & ~exp_CD4\n",
    "adata.obs[\"TRA+_TRB+_CD4+_CD8+\"] = exp_TRA & exp_TRB & exp_CD4 & exp_CD8\n",
    "adata.obs[\"TRA+_TRB+_CD4-_CD8-\"] = exp_TRA & exp_TRB & ~exp_CD4 & ~exp_CD8\n",
    "\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4+_CD8-\"] = (exp_TRA | exp_TRB) & exp_CD4 & ~exp_CD8\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4-_CD8+\"] = (exp_TRA | exp_TRB) & exp_CD8 & ~exp_CD4\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4+_CD8+\"] = (exp_TRA | exp_TRB) & exp_CD4 & exp_CD8\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4-_CD8-\"] = (exp_TRA | exp_TRB) & ~exp_CD4 & ~exp_CD8\n",
    "\n",
    "adata.obs[\"CD45+_(TRA+_or_TRB+)_CD4+_CD8-\"] = exp_PTPRC & (exp_TRA | exp_TRB) & exp_CD4 & ~exp_CD8\n",
    "adata.obs[\"CD45+_(TRA+_or_TRB+)_CD4-_CD8+\"] = exp_PTPRC & (exp_TRA | exp_TRB) & exp_CD8 & ~exp_CD4\n",
    "\n",
    "adata.obs[\"CD3+\"] = exp_CD3\n",
    "adata.obs[\"CD3+_CD4+\"] = exp_CD3 & exp_CD4 \n",
    "adata.obs[\"CD3+_CD8+\"] = exp_CD3 & exp_CD8 \n",
    "adata.obs[\"CD3+_TRA+\"] = exp_CD3 & exp_TRA\n",
    "adata.obs[\"CD3+_TRB+\"] = exp_CD3 & exp_TRB\n",
    "\n",
    "adata.obs[\"CD45+\"] = exp_PTPRC\n",
    "adata.obs[\"CD45+_CD4+\"] = exp_PTPRC & exp_CD4 \n",
    "adata.obs[\"CD45+_CD8+\"] = exp_PTPRC & exp_CD8 \n",
    "adata.obs[\"CD45+_TRA+_TRB-\"] = exp_PTPRC & exp_TRA & ~exp_TRB\n",
    "adata.obs[\"CD45+_TRA-_TRB+\"] = exp_PTPRC & ~exp_TRA & exp_TRB\n",
    "adata.obs[\"CD45+_TRA+_TRB+\"] = exp_PTPRC & exp_TRA & exp_TRB\n",
    "adata.obs[\"CD45+_(TRA+_or_TRB+)\"] = exp_PTPRC & (exp_TRA | exp_TRB)\n",
    "\n",
    "adata.obs[\"IFNG_positive\"] = exp_IFNG\n",
    "\n",
    "adata.obs[\"monocyte_dendritic_cells_macrophages\"] = exp_CSF1R | exp_ZBTB46 | exp_HLA_DRA | exp_LYZ | exp_MS4A6A | exp_H2AB1 | exp_MERTK\n",
    "adata.obs[\"microgria\"] = exp_AIF1 | exp_TMEM119\n",
    "adata.obs[\"astrocyte\"] = exp_SOX9\n",
    "adata.obs[\"neutrophil\"] = exp_CEACAM8\n",
    "adata.obs[\"gd_T_cell\"] = exp_TRGV | exp_TRDV\n",
    "adata.obs[\"gate_out_cell\"] = exp_CSF1R | exp_ZBTB46 | exp_HLA_DRA | exp_LYZ | exp_MS4A6A | exp_H2AB1 | exp_MERTK | exp_AIF1 | exp_TMEM119 | exp_SOX9 | exp_CEACAM8 | exp_TRGV | exp_TRDV\n",
    "\n",
    "adata.obs[\"CD4+_CD8-__gateout\"] = exp_CD4 & ~exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD4-_CD8+__gateout\"] = exp_CD8 & ~exp_CD4 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD4+_CD8+__gateout\"] = exp_CD4 & exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD4-_CD8-__gateout\"] = ~exp_CD4 & ~exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "\n",
    "adata.obs[\"TRA+__gateout\"] = exp_TRA & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRB+__gateout\"] = exp_TRB & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRA+_TRB+__gateout\"] = exp_TRA & exp_TRB & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"(TRA+_or_TRB+)__gateout\"] = exp_TRA | exp_TRB & ~adata.obs[\"gate_out_cell\"]\n",
    "\n",
    "adata.obs[\"TRA+_CD4+__gateout\"] = exp_TRA & exp_CD4 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRA+_CD8+__gateout\"] = exp_TRA & exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRB+_CD4+__gateout\"] = exp_TRB & exp_CD4 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRB+_CD8+__gateout\"] = exp_TRB & exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "\n",
    "adata.obs[\"TRA+_TRB+_CD4+_CD8-__gateout\"] = exp_TRA & exp_TRB & exp_CD4 & ~exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRA+_TRB+_CD4-_CD8+__gateout\"] = exp_TRA & exp_TRB & exp_CD8 & ~exp_CD4 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRA+_TRB+_CD4+_CD8+__gateout\"] = exp_TRA & exp_TRB & exp_CD4 & exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"TRA+_TRB+_CD4-_CD8-__gateout\"] = exp_TRA & exp_TRB & ~exp_CD4 & ~exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4+_CD8-__gateout\"] = (exp_TRA | exp_TRB) & exp_CD4 & ~exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4-_CD8+__gateout\"] = (exp_TRA | exp_TRB) & exp_CD8 & ~exp_CD4 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4+_CD8+__gateout\"] = (exp_TRA | exp_TRB) & exp_CD4 & exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"(TRA+_or_TRB+)_CD4-_CD8-__gateout\"] = (exp_TRA | exp_TRB) & ~exp_CD4 & ~exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "\n",
    "adata.obs[\"CD3+__gateout\"] = exp_CD3 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD3+_CD4+__gateout\"] = exp_CD3 & exp_CD4 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD3+_CD8+__gateout\"] = exp_CD3 & exp_CD8 & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD3+_TRA+__gateout\"] = exp_CD3 & exp_TRA & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD3+_TRB+__gateout\"] = exp_CD3 & exp_TRB & ~adata.obs[\"gate_out_cell\"]\n",
    "\n",
    "adata.obs[\"CD45+__gateout\"] = exp_PTPRC & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD45+_CD4+__gateout\"] = exp_PTPRC & exp_CD4  & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD45+_CD8+__gateout\"] = exp_PTPRC & exp_CD8  & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD45+_TRA+_TRB-__gateout\"] = exp_PTPRC & exp_TRA & ~exp_TRB & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD45+_TRA-_TRB+__gateout\"] = exp_PTPRC & ~exp_TRA & exp_TRB & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD45+_TRA+_TRB+__gateout\"] = exp_PTPRC & exp_TRA & exp_TRB & ~adata.obs[\"gate_out_cell\"]\n",
    "adata.obs[\"CD45+_(TRA+_or_TRB+)__gateout\"] = exp_PTPRC & (exp_TRA | exp_TRB) & ~adata.obs[\"gate_out_cell\"]\n",
    "\n",
    "\n",
    "total_cells = adata.n_obs\n",
    "print(\"total_cells:\", total_cells)\n",
    "\n",
    "for col in adata.obs.columns:\n",
    "    if adata.obs[col].dtype == bool:   \n",
    "        print(f\"{col}: {adata.obs[col].sum()}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c74e629b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffdcbee9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Define subsets for analysis >>>>>>>\n",
    "\n",
    "subset1 = adata.copy()  \n",
    "\n",
    "#subset2 = adata[adata.obs[\"CD4+_CD8-__gateout\"]].copy()\n",
    "\n",
    "#subset3 = adata[adata.obs[\"CD4-_CD8+__gateout\"]].copy()\n",
    "\n",
    "#subset4 = adata[adata.obs[\"CD4+_CD8-\"]].copy()\n",
    "\n",
    "#subset5 = adata[adata.obs[\"CD4-_CD8+\"]].copy()\n",
    "\n",
    "#subset6 = adata[(adata.obs[\"CD4+_CD8-\"]) |(adata.obs[\"CD4-_CD8+\"]) | (adata.obs[\"CD4+_CD8+\"])].copy()\n",
    "\n",
    "#subset7 = adata[adata.obs[\"TRA+\"]].copy()\n",
    "\n",
    "#subset8 = adata[adata.obs[\"TRB+\"]].copy()\n",
    "\n",
    "#subset9 = adata[adata.obs[\"(TRA+_or_TRB+)\"]].copy()\n",
    "\n",
    "subset10 = adata[adata.obs[\"CD45+\"]].copy()\n",
    "\n",
    "\n",
    "\n",
    "subsets = [\n",
    "    (\"human_all_cells\", subset1),\n",
    "#    (\"human_CD4+_CD8-_gateout\", subset2),\n",
    "#    (\"human_CD4-_CD8+_gateout\", subset3),\n",
    "#    (\"human_CD4+_CD8-\", subset4),\n",
    "#    (\"human_CD4-_CD8+\", subset5),\n",
    "#    (\"human_CD4+_or_CD8+\", subset6),\n",
    "#    (\"human_TRA+\", subset7),\n",
    "#    (\"human_TRB+\", subset8),\n",
    "#    (\"human_TRA+_or_TRB+\", subset9),\n",
    "    (\"human_CD45+\", subset10)\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d8b664e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "human_gene_list = [\n",
    "    \"CD3D\", \"CD3E\", \"CD3G\", \"CD4\", \"CD8A\", \"CD8B\", \"CD44\", \"SELL\", \"CCR7\", \"KLF2\", \n",
    "    \"S1PR1\", \"NR4A1\", \"NR4A2\", \"NR4A3\", \"JAK1\", \"JAK2\", \"JAK3\",\"ISG15\", \"ISG20\", \"RORA\", \n",
    "    \"ZNF683\", \"TBX21\", \"TCF7\", \"TOX\", \"GATA3\", \"CXCR3\",\"CXCR6\", \"CD69\", \"IFNG\", \"IL10\", \n",
    "    \"IL21\", \"CXCR5\", \"FOXP3\", \"CTLA4\", \"CX3CR1\", \"GZMA\", \"GZMB\", \"GZMK\", \"PRF1\", \"CCL5\", \n",
    "    \"ITGAE\", \"ITGA1\"\n",
    "]\n",
    "\n",
    "existing_genes = [g for g in human_gene_list if g in adata.var_names]\n",
    "\n",
    "missing_genes = [g for g in human_gene_list if g not in adata.var_names]\n",
    "\n",
    "if missing_genes:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(missing_genes)\n",
    "\n",
    "\n",
    "\n",
    "CD4_list = [\n",
    "    \"ISG15\", \"ISG20\", \"IFI6\", \"IFIT5\", \"IFI35\", \"IFITM1\", \"IRF7\", \"GBP4\", \"CXCL16\"\n",
    "]\n",
    "\n",
    "CD4_genes = [g for g in CD4_list if g in adata.var_names]\n",
    "\n",
    "CD4_missing_genes = [g for g in CD4_list if g not in adata.var_names]\n",
    "\n",
    "if CD4_missing_genes:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(CD4_missing_genes)\n",
    "\n",
    "\n",
    "\n",
    "CD8_list = [\n",
    "    \"ISG15\", \"ISG20\", \"CCL5\", \"IFI44\", \"OAS1\"\n",
    "]\n",
    "\n",
    "CD8_genes = [g for g in CD8_list if g in adata.var_names]\n",
    "\n",
    "CD8_missing_genes = [g for g in CD8_list if g not in adata.var_names]\n",
    "\n",
    "if CD8_missing_genes:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(CD8_missing_genes)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07d56dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< subset analysis >>>>>>>\n",
    "\n",
    "min_cells = 3  \n",
    "for name, subset in subsets:\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs})\")\n",
    "\n",
    "    if subset.n_obs < min_cells:\n",
    "        print(f\"Skip: {name} has {subset.n_obs} cells, which is below {min_cells}.\")\n",
    "        continue\n",
    "\n",
    "    try:\n",
    "\n",
    "        # ======= pre-analysis of subset =======\n",
    "\n",
    "        sc.pp.normalize_total(subset, target_sum=1e4)\n",
    "\n",
    "        subset.raw = subset\n",
    "\n",
    "        sc.pp.log1p(subset)\n",
    "\n",
    "        sc.pp.highly_variable_genes(subset, n_top_genes=2000, subset=False)\n",
    "\n",
    "        if \"highly_variable\" not in subset.var or subset.var[\"highly_variable\"].sum() == 0:\n",
    "            print(f\"Skip: No high-variability genes were detected in {name}\")\n",
    "            continue\n",
    "\n",
    "        subset = subset[:, subset.var[\"highly_variable\"]].copy()\n",
    "\n",
    "        sc.pp.scale(subset, max_value=10)\n",
    "\n",
    "        sc.tl.pca(subset)\n",
    "\n",
    "        sc.pp.neighbors(subset)\n",
    "\n",
    "        sc.tl.umap(subset)\n",
    "\n",
    "        resolution = 1.1\n",
    "\n",
    "        sc.tl.leiden(subset, resolution)\n",
    "\n",
    "        print(subset.X[:5, :5])  \n",
    "        print(subset.var.head())  \n",
    "        print(\"High variable genes:\", subset.var[\"highly_variable\"].sum())\n",
    "        print(subset.obsm.keys())  \n",
    "        print(subset.obsm['X_pca'][:5, :5])  \n",
    "        print(subset.obsm['X_pca'].shape)  \n",
    "        print(subset.uns['neighbors'].keys())\n",
    "        print(subset.obsp['distances'].shape)\n",
    "        print(subset.obsp['connectivities'].shape)\n",
    "        print(subset.obsm.keys())  \n",
    "        print(subset.obsm['X_umap'][:5, :])  \n",
    "        print(subset.obs.keys())  \n",
    "        print(subset.obs['leiden'].value_counts()) \n",
    "\n",
    "        # ======= Cluster =======\n",
    "        print(f\"Processing: {name} - Cluster analysis\")\n",
    "\n",
    "        #UMAP\n",
    "        sc.pl.umap(\n",
    "            subset, \n",
    "            color=\"leiden\", \n",
    "            show=True, \n",
    "            save = f\"_{name}_subset_cluster_{resolution}_UMAP.pdf\"\n",
    "        )\n",
    "\n",
    "        #DEG\n",
    "        sc.tl.rank_genes_groups(\n",
    "            subset, \n",
    "            groupby=\"leiden\", \n",
    "            method=\"wilcoxon\"\n",
    "        )\n",
    "        sc.pl.rank_genes_groups(\n",
    "            subset, \n",
    "            n_genes=25, \n",
    "            sharey=False, \n",
    "            show=False, \n",
    "            save = f\"_{name}_subset_cluster_{resolution}_DEG_fig.pdf\"\n",
    "        )\n",
    "\n",
    "        deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "        groups = deg_result[\"names\"].dtype.names\n",
    "\n",
    "        with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_cluster_{resolution}_DEG_results.xlsx\") as writer:\n",
    "            for group in groups:\n",
    "                df = pd.DataFrame({\n",
    "                    \"gene\": deg_result[\"names\"][group],\n",
    "                    \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                    \"pvals\": deg_result[\"pvals\"][group],\n",
    "                    \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                    \"scores\": deg_result[\"scores\"][group],\n",
    "                })\n",
    "                df.to_excel(writer, sheet_name=f\"cluster_{group}\", index=False)\n",
    "\n",
    "        # UMAP - Expression level \n",
    "        sc.pl.umap(\n",
    "            subset,\n",
    "            color=\"IFNG\",  \n",
    "            cmap=\"Reds\",  \n",
    "            show=True,\n",
    "            save=f\"_{name}_subset_IFNG_expression_UMAP.pdf\"\n",
    "        )\n",
    "\n",
    "\n",
    "        # dotplot - cluster\n",
    "        dp = sc.pl.dotplot(\n",
    "            subset, \n",
    "            var_names=existing_genes, \n",
    "            groupby=\"leiden\", \n",
    "            standard_scale='var', \n",
    "            color_map=\"GnBu\", \n",
    "            show=True, \n",
    "            return_fig=True\n",
    "        )\n",
    "        dp.savefig(f\"{cwd}/figures/{name}_subset_cluster_{resolution}_dotplot.pdf\")\n",
    "        \n",
    "\n",
    "        # ======= selected gene-positive cells vs rest =======\n",
    "        positive_name_list = [\n",
    "            \"TRA+\", \"TRB+\", \"TRA+_TRB+\", \"(TRA+_or_TRB+)\", \"(TRA+_or_TRB+)_CD4-_CD8+\", \"(TRA+_or_TRB+)_CD4+_CD8-\", \"CD4+\", \"CD8+\", \"CD4+_CD8-\", \"CD4-_CD8+\"\n",
    "        ]\n",
    "\n",
    "        for positive in positive_name_list:\n",
    "\n",
    "            #UMAP\n",
    "            min_cells_threshold = 3  \n",
    "            true_count = subset.obs[f\"{positive}\"].sum()\n",
    "            false_count = len(subset.obs[f\"{positive}\"]) - true_count\n",
    "\n",
    "            if true_count < min_cells_threshold:\n",
    "                print(f\"Skipping {positive} due to insufficient positive cells (True: {true_count})\")\n",
    "                continue\n",
    "\n",
    "            if false_count < min_cells_threshold:\n",
    "                print(f\"Skipping {positive} due to insufficient negative cells (False: {false_count})\")\n",
    "                continue\n",
    "\n",
    "            else:\n",
    "                print(f\"True: {true_count}  False: {false_count}\")\n",
    "                sub = subset.copy()\n",
    "                sub.obs[f\"{positive}\"] = pd.Categorical(sub.obs[f\"{positive}\"].astype(str),categories=[\"False\", \"True\"])\n",
    "\n",
    "                order = sub.obs[f\"{positive}\"] == \"True\"\n",
    "                na_cells = sub.obs[~order].index\n",
    "                true_cells = sub.obs[order].index\n",
    "\n",
    "                new_order = na_cells.tolist() + true_cells.tolist()\n",
    "\n",
    "                sub = sub[new_order, :]\n",
    "\n",
    "                sc.pl.umap(\n",
    "                    sub, \n",
    "                    color=f\"{positive}\",\n",
    "                    palette=[\"lightgray\", \"blue\"], \n",
    "                    show=True, \n",
    "                    save=f\"_{name}_subset_{positive}_vs_rest_UMAP.pdf\"\n",
    "                )\n",
    "                \n",
    "                dp = sc.pl.dotplot(\n",
    "                    sub, \n",
    "                    var_names=existing_genes, \n",
    "                    groupby=f\"{positive}\", \n",
    "                    standard_scale='var', \n",
    "                    color_map=\"GnBu\", \n",
    "                    show=True, \n",
    "                    return_fig=True\n",
    "                )\n",
    "                dp.savefig(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_dotplot.pdf\")\n",
    "\n",
    "                dp = sc.pl.dotplot(\n",
    "                    sub, \n",
    "                    var_names=CD4_genes, \n",
    "                    groupby=f\"{positive}\", \n",
    "                    standard_scale='var', \n",
    "                    color_map=\"GnBu\", \n",
    "                    show=True, \n",
    "                    return_fig=True\n",
    "                )\n",
    "                dp.savefig(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_dotplot_CD4.pdf\")\n",
    "\n",
    "                dp = sc.pl.dotplot(\n",
    "                    sub, \n",
    "                    var_names=CD8_genes, \n",
    "                    groupby=f\"{positive}\", \n",
    "                    standard_scale='var', \n",
    "                    color_map=\"GnBu\", \n",
    "                    show=True, \n",
    "                    return_fig=True\n",
    "                )\n",
    "                dp.savefig(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_dotplot_CD8.pdf\")\n",
    "\n",
    "                custom_palette = {\"True\": \"#1f77b4\", \"False\": \"#B0B0B0\"}\n",
    "\n",
    "                sc.pl.violin(\n",
    "                    sub, \n",
    "                    keys=CD4_genes, \n",
    "                    groupby=f\"{positive}\", \n",
    "                    palette=custom_palette, \n",
    "                    stripplot=True, \n",
    "                    jitter=0.25, \n",
    "                    size=2, \n",
    "                    rotation=90, \n",
    "                    show=True, \n",
    "                    save=f\"_{name}_subset_{positive}_vs_rest_violinplot_CD4.pdf\"\n",
    "                )\n",
    "\n",
    "                sc.pl.violin(\n",
    "                    sub, \n",
    "                    keys=CD8_genes, \n",
    "                    groupby=f\"{positive}\", \n",
    "                    palette=custom_palette, \n",
    "                    stripplot=True, \n",
    "                    jitter=0.25, \n",
    "                    size=2, \n",
    "                    rotation=90, \n",
    "                    show=True, \n",
    "                    save=f\"_{name}_subset_{positive}_vs_rest_violinplot_CD8.pdf\"\n",
    "                )\n",
    "\n",
    "                sc.tl.rank_genes_groups(\n",
    "                    sub, \n",
    "                    groupby=f\"{positive}\", \n",
    "                    method=\"wilcoxon\"\n",
    "                )\n",
    "\n",
    "                sc.pl.rank_genes_groups(\n",
    "                    sub, \n",
    "                    n_genes=25, \n",
    "                    sharey=False, \n",
    "                    show=False, \n",
    "                    save=f\"_{name}_subset_{positive}_vs_rest_DEG_fig.pdf\"\n",
    "                )\n",
    "\n",
    "                deg_result = sub.uns[\"rank_genes_groups\"]\n",
    "                groups = deg_result[\"names\"].dtype.names\n",
    "\n",
    "                with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_DEG_results.xlsx\") as writer:\n",
    "                    for group in groups:\n",
    "                        df = pd.DataFrame({\n",
    "                            \"gene\": deg_result[\"names\"][group],\n",
    "                            \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                            \"pvals\": deg_result[\"pvals\"][group],\n",
    "                            \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                            \"scores\": deg_result[\"scores\"][group],\n",
    "                        })\n",
    "                        df.to_excel(writer, sheet_name=f\"cluster_{group}\", index=False)\n",
    "\n",
    "\n",
    "                group = \"True\"\n",
    "                df_deg = pd.DataFrame({\n",
    "                    \"gene\": deg_result[\"names\"][group],\n",
    "                    \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                    \"pvals\": deg_result[\"pvals\"][group],\n",
    "                    \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                    \"scores\": deg_result[\"scores\"][group],\n",
    "                })\n",
    "\n",
    "                df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "                df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "                def plot_volcano(\n",
    "                        df, \n",
    "                        name, \n",
    "                        cwd,\n",
    "                        label_genes=True, \n",
    "                        xlim=None, \n",
    "                        ylim=None, \n",
    "                        label_specific_genes=None, \n",
    "                        label_suffix=None, \n",
    "                        base_ns_color=\"#B0B0B0\", \n",
    "                        base_sig_color=\"#d62728\", \n",
    "                        highlight_color=\"#1f77b4\"\n",
    "                    ):\n",
    "\n",
    "                    df = df.copy()\n",
    "                    if label_specific_genes is not None:\n",
    "                        hi = {g.upper() for g in label_specific_genes}\n",
    "                        df[\"is_highlight\"] = df[\"gene\"].str.upper().isin(hi)\n",
    "                        genes_to_label = df[df[\"is_highlight\"]]\n",
    "                        if label_suffix is None:\n",
    "                            label_suffix = \"specificgenes\"\n",
    "\n",
    "                    elif label_genes:\n",
    "                        df[\"is_highlight\"] = df[\"significant\"]\n",
    "                        genes_to_label = df[df[\"significant\"]]\n",
    "                        if label_suffix is None:\n",
    "                            label_suffix = \"label\"\n",
    "\n",
    "                    else:\n",
    "                        df[\"is_highlight\"] = False\n",
    "                        genes_to_label = pd.DataFrame()\n",
    "                        if label_suffix is None:\n",
    "                            label_suffix = \"nolabel\"\n",
    "\n",
    "                    df[\"_ns\"]  = ~df[\"significant\"]\n",
    "                    df[\"_sig\"] = df[\"significant\"] & ~df[\"is_highlight\"]\n",
    "                    df[\"_hi\"]  = df[\"is_highlight\"]\n",
    "\n",
    "                    def _draw(ax, fix_range=False):\n",
    "                        point_size = 24  \n",
    "\n",
    "                        ax.scatter(\n",
    "                            df.loc[df[\"_ns\"], \"logfoldchanges\"],\n",
    "                            df.loc[df[\"_ns\"], \"-log10_pval\"],\n",
    "                            c=base_ns_color, s=point_size, alpha=0.7, edgecolors=\"none\", zorder=1\n",
    "                        )\n",
    "                        ax.scatter(\n",
    "                            df.loc[df[\"_sig\"], \"logfoldchanges\"],\n",
    "                            df.loc[df[\"_sig\"], \"-log10_pval\"],\n",
    "                            c=base_sig_color, s=point_size, alpha=0.9, edgecolors=\"none\", zorder=2\n",
    "                        )\n",
    "                        ax.scatter(\n",
    "                            df.loc[df[\"_hi\"], \"logfoldchanges\"],\n",
    "                            df.loc[df[\"_hi\"], \"-log10_pval\"],\n",
    "                            c=highlight_color, s=point_size, alpha=1.0, edgecolors=\"none\", zorder=4\n",
    "                        )\n",
    "\n",
    "                        ax.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "                        ax.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "                        ax.axvline( 2, linestyle='--', color='black', linewidth=0.8)\n",
    "\n",
    "                        if not genes_to_label.empty:\n",
    "                            to_annot = genes_to_label.sort_values(\"-log10_pval\", ascending=False)\n",
    "                            offsets = [(10, 10), (10, -10), (-10, 10), (-10, -10)]  \n",
    "                            for i, (_, r) in enumerate(to_annot.iterrows()):\n",
    "                                dx, dy = offsets[i % len(offsets)]\n",
    "                                x0, y0 = r[\"logfoldchanges\"], r[\"-log10_pval\"]\n",
    "                                x1, y1 = x0 + dx * 0.05, y0 + dy * 0.05\n",
    "\n",
    "                                ax.text(\n",
    "                                    x1, y1, r[\"gene\"],\n",
    "                                    fontsize=9, weight=\"bold\", color=\"black\", zorder=6,\n",
    "                                    path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]\n",
    "                                )\n",
    "                                ax.plot([x0, x1], [y0, y1], color=\"black\", linewidth=0.8, zorder=5)\n",
    "\n",
    "                        ax.set_xlabel(\"log2 Fold Change\")\n",
    "                        ax.set_ylabel(\"-log10(p-value)\")\n",
    "                        ax.set_title(f\"{name}_subset ({positive} vs rest)\")\n",
    "\n",
    "                        if fix_range:\n",
    "                            if xlim is not None: ax.set_xlim(xlim)\n",
    "                            if ylim is not None: ax.set_ylim(ylim)\n",
    "                            ax.autoscale(enable=False)\n",
    "                            ax.margins(x=0, y=0)\n",
    "\n",
    "                    plt.figure(figsize=(8, 6))\n",
    "                    ax = plt.gca()\n",
    "                    _draw(ax, fix_range=False)\n",
    "                    plt.tight_layout()\n",
    "                    plt.savefig(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_volcano_plot_{label_suffix}.pdf\")\n",
    "                    plt.close()\n",
    "\n",
    "                    fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "                    _draw(ax, fix_range=True)\n",
    "                    fig.savefig(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_volcano_plot_{label_suffix}_fixedrange.pdf\", bbox_inches=\"tight\")\n",
    "                    plt.close(fig)\n",
    "\n",
    "\n",
    "\n",
    "                plot_volcano(\n",
    "                    df_deg, \n",
    "                    name, \n",
    "                    cwd, \n",
    "                    label_genes=True, \n",
    "                    xlim=[-10, 10]\n",
    "                )\n",
    "\n",
    "                plot_volcano(\n",
    "                    df_deg, \n",
    "                    name, \n",
    "                    cwd, \n",
    "                    label_genes=False, \n",
    "                    xlim=[-10, 10]\n",
    "                )\n",
    "\n",
    "                specificgenesCD4 = [\n",
    "                    \"CD109\", \"TLR3\", \"TRAC\", \"CD4\", \"CCL2\",\"CXCL16\", \"GBP4\", \"NOTCH1\", \"IFITM1\", \"SATB1\",\"IRF7\", \"ISG15\", \"VCAM1\", \"JAG1\", \"SEMA4C\",\"IFI6\", \"IFIT5\", \"IFI35\", \"MX2\", \"IFIH1\"\n",
    "                ]\n",
    "                plot_volcano(\n",
    "                    df_deg,\n",
    "                    name, \n",
    "                    cwd, \n",
    "                    label_genes=False, \n",
    "                    xlim=[-10, 10], \n",
    "                    label_specific_genes=specificgenesCD4, \n",
    "                    label_suffix = \"specificgenesCD4\"\n",
    "                )\n",
    "\n",
    "                specificgenesCD8 = [\n",
    "                    \"CD8B\", \"TRAC\", \"CD8A\", \"SEMA4A\", \"CCL4\", \"CCL5\",\"TOX\", \"IFI44\", \"JAML\", \"OAS1\", \"ISG15\"\n",
    "                ]\n",
    "                plot_volcano(\n",
    "                    df_deg, \n",
    "                    name, \n",
    "                    cwd, \n",
    "                    label_genes=False, \n",
    "                    xlim=[-10, 10], \n",
    "                    label_specific_genes=specificgenesCD8, \n",
    "                    label_suffix = \"specificgenesCD8\"\n",
    "                )\n",
    "\n",
    "\n",
    "        # ===================== TRA+_TRB+ cells vs. cells positive for the specified gene ====================\n",
    "\n",
    "        print(subset.obs[\"TRA+\"].dtype)\n",
    "        print(subset.obs[\"TRA+\"].unique())\n",
    "        print(subset.obs[\"TRB+\"].dtype)\n",
    "        print(subset.obs[\"TRB+\"].unique())\n",
    "\n",
    "        subset.obs[\"TRA+_TRB+\"] = (subset.obs[\"TRA+\"]) & (subset.obs[\"TRB+\"])\n",
    "        subset.obs[\"TRA+_TRB-\"] = (subset.obs[\"TRA+\"]) & ~(subset.obs[\"TRB+\"])\n",
    "        subset.obs[\"TRA-_TRB+\"] = ~(subset.obs[\"TRA+\"]) & (subset.obs[\"TRB+\"])\n",
    "        subset.obs[\"TRA-_TRB-\"] = ~(subset.obs[\"TRA+\"]) & ~(subset.obs[\"TRB+\"])\n",
    "\n",
    "        print(f\"TRA+_TRB+ counts:\\n{subset.obs['TRA+_TRB+'].value_counts()}\")\n",
    "        print(f\"TRA+_TRB- counts:\\n{subset.obs['TRA+_TRB-'].value_counts()}\")\n",
    "        print(f\"TRA-_TRB+ counts:\\n{subset.obs['TRA-_TRB+'].value_counts()}\")\n",
    "        print(f\"TRA-_TRB- counts:\\n{subset.obs['TRA-_TRB-'].value_counts()}\")\n",
    "\n",
    "\n",
    "        subset.obs[\"group\"] = \"other\"  \n",
    "        subset.obs.loc[subset.obs[\"TRA+_TRB+\"], \"group\"] = \"TRA+_TRB+\"\n",
    "        subset.obs.loc[subset.obs[\"TRA+_TRB-\"], \"group\"] = \"TRA+_TRB-\"\n",
    "        subset.obs.loc[subset.obs[\"TRA-_TRB+\"], \"group\"] = \"TRA-_TRB+\"\n",
    "        subset.obs.loc[subset.obs[\"TRA-_TRB-\"], \"group\"] = \"TRA-_TRB-\"\n",
    "\n",
    "        groupname_list = [\"TRA+_TRB-\", \"TRA-_TRB+\", \"TRA-_TRB-\"]\n",
    "\n",
    "        for groupname in groupname_list:\n",
    "\n",
    "            min_cells_threshold = 3\n",
    "            group_counts = subset.obs[\"group\"].value_counts()\n",
    "            if group_counts.get(\"TRA+_TRB+\", 0) < min_cells_threshold:\n",
    "                print(f\"Skipping due to insufficient positive cells (TRA+_TRB+: {group_counts.get('TRA+_TRB+', 0)})\")\n",
    "                continue  \n",
    "\n",
    "            if group_counts.get(groupname, 0) < min_cells_threshold:\n",
    "                print(f\"Skipping due to insufficient negative cells ({groupname}: {group_counts.get(groupname, 0)})\")\n",
    "                continue  \n",
    "            else:\n",
    "\n",
    "                subset = subset[subset.obs[\"group\"].sort_values().index, :]\n",
    "                sc.pl.umap(\n",
    "                    subset, \n",
    "                    color=\"group\", \n",
    "                    palette=[\"red\", \"blue\", \"green\", \"lightgray\"], \n",
    "                    show=True, \n",
    "                    save=f\"_{name}_subset_TRA+_TRB+_vs_{groupname}_UMAP.pdf\"\n",
    "                )\n",
    "\n",
    "                subset_filtered = subset[subset.obs[\"group\"].isin([\"TRA+_TRB+\", groupname]), :]\n",
    "\n",
    "\n",
    "                sc.tl.rank_genes_groups(\n",
    "                    subset_filtered, \n",
    "                    groupby=\"group\", \n",
    "                    groups=[\"TRA+_TRB+\", groupname], \n",
    "                    method=\"wilcoxon\"\n",
    "                )\n",
    "                sc.pl.rank_genes_groups(\n",
    "                    subset_filtered, \n",
    "                    n_genes=25, \n",
    "                    sharey=False, \n",
    "                    show=False, \n",
    "                    save=f\"_{name}_subset_TRA+_TRB+_vs_{groupname}_DEG_fig.pdf\"\n",
    "                )\n",
    "\n",
    "                deg_result = subset_filtered.uns[\"rank_genes_groups\"]\n",
    "                groups = deg_result[\"names\"].dtype.names\n",
    "                with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_TRA+_TRB+_vs_{groupname}_DEG_results.xlsx\") as writer:\n",
    "                    for group in groups:\n",
    "                        df = pd.DataFrame({\n",
    "                            \"gene\": deg_result[\"names\"][group],\n",
    "                            \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                            \"pvals\": deg_result[\"pvals\"][group],\n",
    "                            \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                            \"scores\": deg_result[\"scores\"][group],\n",
    "                        })\n",
    "                        df.to_excel(writer, sheet_name=f\"{group[:31]}\", index=False)\n",
    "\n",
    "\n",
    "                # --- dotplot ---\n",
    "                dp = sc.pl.dotplot(\n",
    "                    subset_filtered, \n",
    "                    var_names=existing_genes, \n",
    "                    groupby=\"group\", \n",
    "                    standard_scale='var', \n",
    "                    color_map=\"GnBu\", \n",
    "                    show=True, \n",
    "                    return_fig=True\n",
    "                )\n",
    "                dp.savefig(f\"{cwd}/figures/{name}_subset_TRA+_TRB+_vs_{groupname}_dotplot.pdf\")\n",
    "\n",
    "\n",
    "                # --- volcano plot ---\n",
    "                deg_result = subset_filtered.uns[\"rank_genes_groups\"]\n",
    "                group = \"TRA+_TRB+\"\n",
    "                df_deg = pd.DataFrame({\n",
    "                    \"gene\": deg_result[\"names\"][group],\n",
    "                    \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                    \"pvals\": deg_result[\"pvals\"][group],\n",
    "                    \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                    \"scores\": deg_result[\"scores\"][group],\n",
    "                })\n",
    "                \n",
    "                exclude_genes = [\"TRAC\", \"TRBC1\", \"TRBC2\"]\n",
    "                df_deg = df_deg[~df_deg[\"gene\"].isin(exclude_genes)].copy()\n",
    "\n",
    "                df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "                df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "                def plot_volcano(\n",
    "                        df, \n",
    "                        name, \n",
    "                        cwd, \n",
    "                        label_genes=True, \n",
    "                        xlim=[-10, 10], \n",
    "                        ylim=None\n",
    "                    ):\n",
    "\n",
    "                    plt.figure(figsize=(8, 6))\n",
    "                    plt.scatter(\n",
    "                        df[\"logfoldchanges\"], df[\"-log10_pval\"],\n",
    "                        c=df[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "                        alpha=0.7, edgecolors=\"none\"\n",
    "                    )\n",
    "                    if label_genes:\n",
    "                        for _, row in df[df[\"significant\"]].iterrows():\n",
    "                            plt.text(\n",
    "                                row[\"logfoldchanges\"],\n",
    "                                row[\"-log10_pval\"],\n",
    "                                row[\"gene\"],\n",
    "                                fontsize=8,\n",
    "                                ha='center',\n",
    "                                va='bottom'\n",
    "                            )\n",
    "                    plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)  \n",
    "                    plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)  \n",
    "                    plt.axvline(2, linestyle='--', color='black', linewidth=0.8)   \n",
    "                    plt.xlabel(\"log2 Fold Change\")\n",
    "                    plt.ylabel(\"-log10(p-value)\")\n",
    "                    plt.title(f\"{name}_subset ({positive} vs rest)\")\n",
    "                    plt.tight_layout()\n",
    "\n",
    "                    label_suffix = \"label\" if label_genes else \"nolabel\"\n",
    "                    plt.savefig(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_volcano_plot_{label_suffix}.pdf\")\n",
    "                    plt.close()\n",
    "\n",
    "                    fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "\n",
    "                    plt.scatter(\n",
    "                        df[\"logfoldchanges\"], df[\"-log10_pval\"],\n",
    "                        c=df[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "                        alpha=0.7, edgecolors=\"none\"\n",
    "                    )\n",
    "                    if label_genes:\n",
    "                        for _, row in df[df[\"significant\"]].iterrows():\n",
    "                            plt.text(\n",
    "                                row[\"logfoldchanges\"],\n",
    "                                row[\"-log10_pval\"],\n",
    "                                row[\"gene\"],\n",
    "                                fontsize=8,\n",
    "                                ha='center',\n",
    "                                va='bottom',\n",
    "                                clip_on=True\n",
    "                            )\n",
    "\n",
    "                    plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "                    plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "                    plt.axvline(2, linestyle='--', color='black', linewidth=0.8)\n",
    "                    plt.xlabel(\"log2 Fold Change\")\n",
    "                    plt.ylabel(\"-log10(p-value)\")\n",
    "                    plt.title(f\"{name}_subset ({positive} vs rest)\")\n",
    "\n",
    "                    if xlim is not None:\n",
    "                        plt.xlim(xlim)\n",
    "                    if ylim is not None:\n",
    "                        plt.ylim(ylim)\n",
    "                    ax.autoscale(enable=False)  \n",
    "                    ax.margins(x=0, y=0) \n",
    "\n",
    "                    plt.savefig(f\"{cwd}/figures/{name}_subset_{positive}_vs_rest_volcano_plot_{label_suffix}_fixedrange.pdf\", bbox_inches=\"tight\")\n",
    "                    plt.close()\n",
    "\n",
    "                plot_volcano(df_deg, name, cwd, label_genes=True)\n",
    "                plot_volcano(df_deg, name, cwd, label_genes=False)\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error in {name}: {e}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd244e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< gene expression dotplot (subset) >>>>>>>\n",
    "\n",
    "existing_genes = [g for g in human_gene_list if g in adata.var_names]\n",
    "\n",
    "missing_genes = [g for g in human_gene_list if g not in adata.var_names]\n",
    "\n",
    "if missing_genes:\n",
    "    print(\"The following genes were not found (excluded):\")\n",
    "    print(missing_genes)\n",
    "\n",
    "subsets_of_a_cell = []\n",
    "\n",
    "for name, subset in subsets:\n",
    "    subset_cp = subset.copy()\n",
    "    subset_cp.obs[\"subset_name\"] = name\n",
    "    subset_cp.obs_names = [f\"{name}_{i}\" for i in subset_cp.obs_names]  \n",
    "    subsets_of_a_cell.append(subset_cp)\n",
    "\n",
    "adata_subsets_of_a_cell = ad.concat(subsets_of_a_cell, join=\"outer\")\n",
    "\n",
    "dp = sc.pl.dotplot(\n",
    "    adata_subsets_of_a_cell,\n",
    "    var_names=existing_genes,\n",
    "    groupby=\"subset_name\",\n",
    "    standard_scale='var',\n",
    "    color_map=\"GnBu\",\n",
    "    show=True,\n",
    "    return_fig=True\n",
    ")\n",
    "\n",
    "dp.savefig(f\"{cwd}/figures/dotplot_by_subset.pdf\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4c35678",
   "metadata": {},
   "source": [
    "# Rename in Tarminal\n",
    "\n",
    "for f in *mouse*; do\n",
    "  new=$(echo \"$f\" | sed 's/^.*mouse/mouse/')\n",
    "  mv \"$f\" \"$new\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5edad8e3",
   "metadata": {},
   "source": [
    "# Rename in PowerShell\n",
    "\n",
    "Get-ChildItem -File *mouse* | ForEach-Object {\n",
    "    $newName = $_.Name -replace '.*mouse','mouse'\n",
    "    if ($_.Name -ne $newName) {\n",
    "        Rename-Item -LiteralPath $_.FullName -NewName $newName\n",
    "        Write-Output \"$($_.Name) → $newName\"\n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Aomine_scRNA-seq_0718",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
