{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e0fb4ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Setting >>>>>>>\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import gseapy\n",
    "import matplotlib.pyplot as plt\n",
    "import scvi\n",
    "import scipy.sparse\n",
    "import math\n",
    "import os\n",
    "import anndata\n",
    "import matplotlib.patheffects as pe\n",
    "\n",
    "sc.settings.verbosity = 3 \n",
    "\n",
    "print(os.getcwd())\n",
    "cwd = f\"{os.getcwd()}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8183e30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Integrating scRNA-seq and scTCR-seq Data >>>>>>>\n",
    "\n",
    "batch_list = [\"Control_SP\", \"Control_TG\"]  \n",
    "\n",
    "for batch in batch_list:\n",
    "\n",
    "    tcr_path = f\"{cwd}/filtered_contig_annotations_{batch}_TCR.csv\"\n",
    "    h5_path = f\"{cwd}/filtered_feature_bc_matrix_{batch}.h5\"\n",
    "\n",
    "    adata = sc.read_10x_h5(h5_path)\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"batch\"] = batch\n",
    "\n",
    "    tcr_df = pd.read_csv(tcr_path)\n",
    "    tcr_df = tcr_df[tcr_df[\"high_confidence\"] & tcr_df[\"productive\"]]\n",
    "    tcr_df[\"barcode\"] = tcr_df[\"barcode\"].str.replace(\"-1\", \"\", regex=False)\n",
    "    tcr_df[\"joint_barcode\"] = tcr_df[\"barcode\"] + \"-\" + batch\n",
    "\n",
    "    tra_df = (\n",
    "        tcr_df[tcr_df[\"chain\"] == \"TRA\"]\n",
    "        .drop_duplicates(subset=\"joint_barcode\", keep=\"first\")\n",
    "        [[\"joint_barcode\", \"v_gene\", \"j_gene\", \"cdr3\"]]\n",
    "        .rename(columns={\"v_gene\": \"v_gene_TRA\", \"j_gene\": \"j_gene_TRA\", \"cdr3\": \"cdr3_TRA\"})\n",
    "    )\n",
    "\n",
    "    trb_df = (\n",
    "        tcr_df[tcr_df[\"chain\"] == \"TRB\"]\n",
    "        .drop_duplicates(subset=\"joint_barcode\", keep=\"first\")\n",
    "        [[\"joint_barcode\", \"v_gene\", \"j_gene\", \"cdr3\"]]\n",
    "        .rename(columns={\"v_gene\": \"v_gene_TRB\", \"j_gene\": \"j_gene_TRB\", \"cdr3\": \"cdr3_TRB\"})\n",
    "    )\n",
    "\n",
    "    tcr_merged = pd.merge(tra_df, trb_df, on=\"joint_barcode\", how=\"outer\")\n",
    "\n",
    "    adata.obs.index = adata.obs.index.str.replace(\"-1\", \"\", regex=False)\n",
    "    adata.obs[\"joint_barcode\"] = adata.obs.index + \"-\" + batch\n",
    "    adata.obs = adata.obs.set_index(\"joint_barcode\")\n",
    "    adata.obs = adata.obs.join(tcr_merged.set_index(\"joint_barcode\"), how=\"left\")\n",
    "\n",
    "    print(\"Number of cells with TRB information::\", adata.obs[\"cdr3_TRB\"].notna().sum())\n",
    "    print(adata.obs[[\"v_gene_TRA\", \"cdr3_TRA\", \"v_gene_TRB\", \"cdr3_TRB\"]].dropna().head())\n",
    "\n",
    "    adata.write(f\"{cwd}/adata_combined_TCR_{batch}_raw.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "523d1c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Merging multiple AnnData objects across batches >>>>>>>\n",
    "\n",
    "samples = [\n",
    "    {\"path\": f\"{cwd}/adata_combined_TCR_Control_SP_raw.h5ad\", \"batch\": \"Control_SP\"},\n",
    "    {\"path\": f\"{cwd}/adata_combined_TCR_Control_TG_raw.h5ad\", \"batch\": \"Control_TG\"},\n",
    "]\n",
    "\n",
    "adatas = []\n",
    "for sample in samples:\n",
    "    ad = sc.read_h5ad(sample[\"path\"])\n",
    "    ad.obs[\"batch\"] = sample[\"batch\"]  \n",
    "    adatas.append(ad)\n",
    "\n",
    "adata_combined = adatas[0].concatenate(adatas[1:], batch_key=\"batch_group\", index_unique=None)\n",
    "\n",
    "print(adata_combined.shape)\n",
    "print(adata_combined.obs[\"batch\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e28a1b48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<<  Save (Optional) >>>>>>>\n",
    "\n",
    "adata_combined.write(f\"{cwd}/adata_combined_TCR_all_batchs_raw.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15644fc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Concatenate CDR3,V,J (TRA + TRB) >>>>>>>\n",
    "\n",
    "adata_combined.obs[\"cdr3_TRA_TRB\"] = adata_combined.obs[\"cdr3_TRA\"].astype(str) + \"_\" + adata_combined.obs[\"cdr3_TRB\"].astype(str)\n",
    "\n",
    "adata_combined.obs[\"V_TRA_TRB\"] = adata_combined.obs[\"v_gene_TRA\"].astype(str) + \"_\" + adata_combined.obs[\"v_gene_TRB\"].astype(str)\n",
    "\n",
    "adata_combined.obs[\"J_TRA_TRB\"] = adata_combined.obs[\"j_gene_TRA\"].astype(str) + \"_\" + adata_combined.obs[\"j_gene_TRB\"].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18a35b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<<  filter >>>>>>>\n",
    "\n",
    "sc.pp.filter_cells(adata_combined, min_genes=200)\n",
    "sc.pp.filter_genes(adata_combined, min_cells=3)\n",
    "\n",
    "adata_combined.raw = adata_combined.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "688d39ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Save  >>>>>>>\n",
    "\n",
    "adata_combined.write(f\"{cwd}/adata_combined_TCR_all_batchs.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb8ecc00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Reload >>>>>>>\n",
    "\n",
    "adata = sc.read_h5ad(f\"{cwd}/adata_combined_TCR_all_batchs.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3633086",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Create labels by gene expression >>>>>>>\n",
    "\n",
    "def get_expr_or_zero(adata, gene):\n",
    "    if gene in adata.var_names:\n",
    "        x = adata[:, gene].X\n",
    "        expr = x.toarray().flatten() if not isinstance(x, np.ndarray) else x.flatten()\n",
    "        return expr > 0\n",
    "    else:\n",
    "        print(f\"[skip] {gene} is not exist\")\n",
    "        return np.zeros(adata.n_obs, dtype=bool)\n",
    "    \n",
    "def get_multi_gene_expr(adata, genes):\n",
    "    return np.logical_or.reduce([get_expr_or_zero(adata, g) for g in genes])\n",
    "\n",
    "\n",
    "exp_TRA = get_multi_gene_expr(adata, [\"Trac\"]) > 0\n",
    "exp_TRB = (\n",
    "    (get_multi_gene_expr(adata, [\"Trbc1\"]) > 0) | \n",
    "    (get_multi_gene_expr(adata, [\"Trbc2\"]) > 0)\n",
    ")\n",
    "exp_TRB1 = get_multi_gene_expr(adata, [\"Trbc1\"]) > 0\n",
    "exp_TRB2 = get_multi_gene_expr(adata, [\"Trbc2\"]) > 0\n",
    "exp_TRAB = (\n",
    "    (get_multi_gene_expr(adata, [\"Trac\"]) > 0 ) &\n",
    "    (get_multi_gene_expr(adata, [\"Trbc1\"]) > 0) | \n",
    "    (get_multi_gene_expr(adata_combined, [\"Trbc2\"]) > 0)\n",
    ")\n",
    "\n",
    "exp_CD4 = get_multi_gene_expr(adata, [\"Cd4\"]) > 0\n",
    "exp_CD8 = (\n",
    "    (get_multi_gene_expr(adata, [\"Cd8a\"]) > 0) | \n",
    "    (get_multi_gene_expr(adata, [\"Cd8b1\"]) > 0)\n",
    ")\n",
    "exp_CD8a = get_multi_gene_expr(adata, [\"Cd8a\"]) > 0\n",
    "exp_CD8b1 = get_multi_gene_expr(adata, [\"Cd8b1\"]) > 0\n",
    "\n",
    "exp_Csf1r = get_multi_gene_expr(adata, [\"Csf1r\"])\n",
    "exp_Zbtb46 = get_multi_gene_expr(adata, [\"Zbtb46\"])\n",
    "exp_H2_Aa = get_multi_gene_expr(adata, [\"H2-Aa\"])\n",
    "exp_Lyz2 = get_multi_gene_expr(adata, [\"Lyz2\"])\n",
    "exp_Ms4a6c = get_multi_gene_expr(adata, [\"Ms4a6c\"]) \n",
    "exp_Aif1 = get_multi_gene_expr(adata, [\"Aif1\"])\n",
    "exp_Tmem119 = get_multi_gene_expr(adata, [\"Tmem119\"])\n",
    "exp_Sox9 = get_multi_gene_expr(adata, [\"Sox9\"])\n",
    "\n",
    "exp_Ly6g = get_multi_gene_expr(adata, [\n",
    "    \"Ly6g6g\", \"Ly6g\", \"Ly6g6d\", \"Ly6g6e\"\n",
    "])\n",
    "\n",
    "exp_Trgv_1_7 = get_multi_gene_expr(adata, [\n",
    "    \"Trgv1\", \"Trgv2\", \"Trgv3\", \"Trgv4\", \"Trgv5\", \"Trgv6\", \"Trgv7\"\n",
    "])\n",
    "\n",
    "exp_Trdv = get_multi_gene_expr(adata, [\n",
    "    \"Trdv1\", \"Trdv2-1\", \"Trdv2-2\", \"Trdv3\", \"Trdv4\", \"Trdv5\"\n",
    "])\n",
    "\n",
    "\n",
    "adata.obs[\"TRA_TRB_positive\"] = exp_TRA & exp_TRB\n",
    "adata.obs[\"TRA_TRB_positive_CD4_positive\"] = exp_TRA & exp_TRB & exp_CD4\n",
    "adata.obs[\"TRA_TRB_positive_CD8_positive\"] = exp_TRA & exp_TRB & exp_CD8\n",
    "adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"] = exp_TRA & exp_TRB & exp_CD4 & exp_CD8\n",
    "adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"] = exp_TRA & exp_TRB & ~exp_CD4 & ~exp_CD8\n",
    "\n",
    "adata.obs[\"monocyte_dendritic_cells_macrophages\"] = exp_Csf1r | exp_Zbtb46 | exp_H2_Aa | exp_Lyz2 | exp_Ms4a6c \n",
    "adata.obs[\"microgria\"] = exp_Aif1 | exp_Tmem119\n",
    "adata.obs[\"astrocyte\"] = exp_Sox9\n",
    "adata.obs[\"neutrophil\"] = exp_Ly6g\n",
    "adata.obs[\"gd_T_cell\"] = exp_Trgv_1_7 | exp_Trdv\n",
    "adata.obs[\"gate_out_cell\"] = exp_Csf1r | exp_Zbtb46 | exp_H2_Aa | exp_Lyz2 | exp_Ms4a6c | exp_Aif1 | exp_Tmem119 | exp_Sox9 | exp_Ly6g | exp_Trgv_1_7 | exp_Trdv \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e790cdd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Cell count >>>>>>>\n",
    "\n",
    "C1 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C2 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C3 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "C4 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C5 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C6 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C7 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C8 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C9 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "C10 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C11 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C12 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C13 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C14 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C15 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "C16 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C17 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C18 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C19 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C20 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C21 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "C22 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C23 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C24 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C25 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C26 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C27 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "C28 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C29 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C30 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C31 = adata[\n",
    "    (adata.obs[\"monocyte_dendritic_cells_macrophages\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C32 = adata[\n",
    "    (adata.obs[\"monocyte_dendritic_cells_macrophages\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C33 = adata[\n",
    "    (adata.obs[\"monocyte_dendritic_cells_macrophages\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C34 = adata[\n",
    "    (adata.obs[\"microgria\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C35 = adata[\n",
    "    (adata.obs[\"microgria\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C36 = adata[\n",
    "    (adata.obs[\"microgria\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C37 = adata[\n",
    "    (adata.obs[\"astrocyte\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C38 = adata[\n",
    "    (adata.obs[\"astrocyte\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C39 = adata[\n",
    "    (adata.obs[\"astrocyte\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C40 = adata[\n",
    "    (adata.obs[\"neutrophil\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C41 = adata[\n",
    "    (adata.obs[\"neutrophil\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C42 = adata[\n",
    "    (adata.obs[\"neutrophil\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C43 = adata[\n",
    "    (adata.obs[\"gd_T_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C44 = adata[\n",
    "    (adata.obs[\"gd_T_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C45 = adata[\n",
    "    (adata.obs[\"gd_T_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C46 = adata[\n",
    "    (adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C47 = adata[\n",
    "    (adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C48 = adata[\n",
    "    (adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C49 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C50 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) &  \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C51 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()    \n",
    "\n",
    "C52 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"] | adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C53 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"] | adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C54 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"] | adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "# ==========\n",
    "\n",
    "C55 = adata[\n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "C56 = adata[\n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "C57 = adata[\n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "subsets = [\n",
    "    (\"TRA+_TRB+_inSP\", C1),\n",
    "    (\"TRA+_TRB+_inTG\", C2),\n",
    "    (\"TRA+_TRB+_inSPTG\", C3),\n",
    "    (\"TRA+_TRB+_gateout_inSP\", C4),\n",
    "    (\"TRA+_TRB+_gateout_inTG\", C5),\n",
    "    (\"TRA+_TRB+_gateout_inSPTG\", C6),\n",
    "\n",
    "    (\"TRA+_TRB+_CD4+_inSP\", C7),\n",
    "    (\"TRA+_TRB+_CD4+_inTG\", C8),\n",
    "    (\"TRA+_TRB+_CD4+_inSPTG\", C9),\n",
    "    (\"TRA+_TRB+_CD4+_gateout_inSP\", C10),\n",
    "    (\"TRA+_TRB+_CD4+_gateout_inTG\", C11),\n",
    "    (\"TRA+_TRB+_CD4+_gateout_inSPTG\", C12),\n",
    "\n",
    "    (\"TRA+_TRB+_CD8+_inSP\", C13),\n",
    "    (\"TRA+_TRB+_CD8+_inTG\", C14),\n",
    "    (\"TRA+_TRB+_CD8+_inSPTG\", C15),\n",
    "    (\"TRA+_TRB+_CD8+_gateout_inSP\", C16),\n",
    "    (\"TRA+_TRB+_CD8+_gateout_inTG\", C17),\n",
    "    (\"TRA+_TRB+_CD8+_gateout_inSPTG\", C8),\n",
    "\n",
    "    (\"TRA+_TRB+_CD4+_CD8+_inSP\", C19),\n",
    "    (\"TRA+_TRB+_CD4+_CD8+_inTG\", C20),\n",
    "    (\"TRA+_TRB+_CD4+_CD8+_inSPTG\", C21),\n",
    "    (\"TRA+_TRB+_CD4+_CD8+_gateout_inSP\", C22),\n",
    "    (\"TRA+_TRB+_CD4+_CD8+_gateout_inTG\", C23),\n",
    "    (\"TRA+_TRB+_CD4+_CD8+_gateout_inSPTG\", C24),\n",
    "\n",
    "    (\"TRA+_TRB+_CD4-_CD8-_inSP\", C25),\n",
    "    (\"TRA+_TRB+_CD4-_CD8-_inTG\", C26),\n",
    "    (\"TRA+_TRB+_CD4-_CD8-_inSPTG\", C27),\n",
    "    (\"TRA+_TRB+_CD4-_CD8-_gateout_inSP\", C28),\n",
    "    (\"TRA+_TRB+_CD4-_CD8-_gateout_inTG\", C29),\n",
    "    (\"TRA+_TRB+_CD4-_CD8-_gateout_inSPTG\", C30),\n",
    "\n",
    "    (\"monocyte_dendritic_cells_macrophages_inSP\", C31),\n",
    "    (\"monocyte_dendritic_cells_macrophages_inTG\", C32),\n",
    "    (\"monocyte_dendritic_cells_macrophages__inSPTG\", C33),\n",
    "\n",
    "    (\"microgria_inSP\", C34),\n",
    "    (\"microgria_inTG\", C35),\n",
    "    (\"microgria_inSPTG\", C36),\n",
    "\n",
    "    (\"astrocyte_inSP\", C37),\n",
    "    (\"astrocyte_inTG\", C38),\n",
    "    (\"astrocyte_inSPTG\", C39),\n",
    "\n",
    "    (\"neutrophil_inSP\", C40),\n",
    "    (\"neutrophil_inTG\", C41),\n",
    "    (\"neutrophil_inSPTG\", C42),\n",
    "    \n",
    "    (\"gd_T_cell_inSP\", C43),\n",
    "    (\"gd_T_cell_inTG\", C44),\n",
    "    (\"gd_T_cell_inSPTG\", C45),\n",
    "    \n",
    "    (\"gate_out_cell_inSP\", C46),\n",
    "    (\"gate_out_cell_inTG\", C47),\n",
    "    (\"gate_out_cell_inSPTG\", C48),\n",
    "\n",
    "    (\"TRA+_TRB+_CD4+_CD8-_gateout_inSP\", C49),\n",
    "    (\"TRA+_TRB+_CD4+_CD8-_gateout_inTG\", C50),\n",
    "    (\"TRA+_TRB+_CD4+_CD8-_gateout_inSPTG\", C51),\n",
    "    (\"TRA+_TRB+_CD8+DP+DN_gateout_inSP\", C52),\n",
    "    (\"TRA+_TRB+_CD8+DP+DN_gateout_inTG\", C53),\n",
    "    (\"TRA+_TRB+_CD8+DP+DN_gateout_inSPTG\", C54),\n",
    "\n",
    "    (\"all_cells_inSP\", C55),\n",
    "    (\"all_cells_inTG\", C56),\n",
    "    (\"all_cells_inSPTG\", C57)\n",
    "]\n",
    "\n",
    "\n",
    "\n",
    "results = []\n",
    "for name, subset in subsets:\n",
    "    print(f\"{name}:  {subset.n_obs}\")\n",
    "    results.append({\"Subset\": name, \"Cell_Count\": subset.n_obs})\n",
    "\n",
    "df_counts = pd.DataFrame(results)\n",
    "df_counts.to_csv(\"cell_counts_summary.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0b60520",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Define subsets for analysis >>>>>>>\n",
    "\n",
    "subset1 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "subset2 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) &  \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "subset3 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD4_positive\"]) & \n",
    "    (~adata.obs[\"TRA_TRB_positive_CD4_CD8_double_positive\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()    \n",
    "\n",
    "\n",
    "subset4 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"] | adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_TG\")\n",
    "].copy()\n",
    "\n",
    "subset5 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"] | adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    (adata.obs[\"batch\"] == \"Control_SP\")\n",
    "].copy()\n",
    "\n",
    "subset6 = adata[\n",
    "    (adata.obs[\"TRA_TRB_positive_CD8_positive\"] | adata.obs[\"TRA_TRB_positive_CD4_CD8_double_negative\"]) & \n",
    "    (~adata.obs[\"gate_out_cell\"]) & \n",
    "    ((adata.obs[\"batch\"] == \"Control_TG\") | (adata.obs[\"batch\"] == \"Control_SP\"))\n",
    "].copy()\n",
    " \n",
    "\n",
    "\n",
    "subsets = [\n",
    "    (\"mouse_CD4_TG\", subset1),\n",
    "    (\"mouse_CD4_SP\", subset2),\n",
    "    (\"mouse_CD4_TG_SP\", subset3),\n",
    "    (\"mouse_CD8+DP+DN_TG\", subset4),\n",
    "    (\"mouse_CD8+DP+DN_SP\", subset5),\n",
    "    (\"mouse_CD8+DP+DN_TG_SP\", subset6)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24a5b490",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Define gene lists for analysis >>>>>>>\n",
    "\n",
    "gene_list = [\n",
    "    \"Cd3d\", \"Cd3e\", \"Cd3g\", \"Cd4\", \"Cd8a\", \"Cd8b1\", \"Cd44\", \"Sell\", \"Ccr7\",\n",
    "    \"Klf2\", \"S1pr1\", \"Nr4a1\", \"Nr4a2\", \"Nr4a3\", \"Jak1\", \"Jak2\", \"Jak3\",\n",
    "    \"Isg15\", \"Isg20\", \"Rora\", \"Tbx21\", \"Tcf7\", \"Tox\", \"Gata3\", \"Cxcr3\",\n",
    "    \"Cxcr6\", \"Cd69\", \"Il17a\", \"Ifng\", \"Il4\", \"Il10\", \"Il21\", \"Cxcr5\",\n",
    "    \"Pdcd1\", \"Foxp3\", \"Ctla4\", \"Cx3cr1\", \"Gzma\", \"Gzmb\", \"Gzmk\", \"Prf1\",\n",
    "    \"Ccl5\", \"Itgae\", \"Itga1\"\n",
    "]\n",
    "\n",
    "existing_genes = [g for g in gene_list if g in adata.var_names]\n",
    "missing_genes = [g for g in gene_list if g not in adata.var_names]\n",
    "if missing_genes:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(missing_genes)\n",
    "\n",
    "\n",
    "\n",
    "gene_list_2 = [\n",
    "    \"Cd3d\", \"Cd3e\", \"Cd3g\", \"Cd4\", \"Cd8a\", \"Cd8b1\", \"Cd44\", \"Sell\", \"Ccr7\",\n",
    "    \"Klf2\", \"S1pr1\", \"Nr4a1\", \"Nr4a2\", \"Nr4a3\", \"Jak1\", \"Jak2\", \"Jak3\",\n",
    "    \"Isg15\", \"Isg20\", \"Rora\", \"Zfp683\", \"Tbx21\", \"Tcf7\", \"Tox\", \"Gata3\", \"Cxcr3\",\n",
    "    \"Cxcr6\", \"Cd69\", \"Il17a\", \"Ifng\", \"Il10\", \"Il21\", \"Cxcr5\",\n",
    "    \"Pdcd1\", \"Foxp3\", \"Ctla4\", \"Cx3cr1\", \"Gzma\", \"Gzmb\", \"Gzmk\", \"Prf1\",\n",
    "    \"Ccl5\", \"Itgae\", \"Itga1\"\n",
    "]\n",
    "\n",
    "existing_genes_2 = [g for g in gene_list_2 if g in adata.var_names]\n",
    "missing_genes_2 = [g for g in gene_list_2 if g not in adata.var_names]\n",
    "if missing_genes_2:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(missing_genes_2)\n",
    "\n",
    "\n",
    "\n",
    "genes_list_CD4TGSP = [\n",
    "\"Isg15\", \"Isg20\", \"Gbp6\", \"Gbp9\", \"Ccl5\", \"Ifi27l2a\", \"Irf7\", \"Oas1a\"       #\"Cxcl10\"\n",
    "]   \n",
    "\n",
    "existing_genes_CD4TGSP = [g for g in genes_list_CD4TGSP if g in adata.var_names]\n",
    "missing_genes_CD4TGSP = [g for g in genes_list_CD4TGSP if g not in adata.var_names]\n",
    "if missing_genes_CD4TGSP:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(missing_genes_CD4TGSP)\n",
    "\n",
    "\n",
    "\n",
    "genes_list_CD4TG = [\n",
    "\"Oas1a\", \"Ifi27l2a\", \"Gbp6\", \"Gbp9\", \"Ifi47\",\"Zbp1\",  \"Irf7\", \"Isg15\", \"Parp14\"\n",
    "]\n",
    "\n",
    "existing_genes_CD4TG = [g for g in genes_list_CD4TG if g in adata.var_names]\n",
    "missing_genes_CD4TG = [g for g in genes_list_CD4TG if g not in adata.var_names]\n",
    "if missing_genes_CD4TG:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(missing_genes_CD4TG)\n",
    "\n",
    "\n",
    "\n",
    "genes_list_CD8TGSP = [\n",
    "\"Ccl5\", \"Ifitm1\", \"Ifitm2\", \"Isg15\", \"Isg20\", \"Ifitm3\"\n",
    "]\n",
    "\n",
    "existing_genes_CD8TGSP = [g for g in genes_list_CD8TGSP if g in adata.var_names]\n",
    "missing_genes_CD8TGSP = [g for g in genes_list_CD8TGSP if g not in adata.var_names]\n",
    "if missing_genes_CD8TGSP:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(missing_genes_CD8TGSP)\n",
    "\n",
    "\n",
    "\n",
    "genes_list_CD8TG = [\n",
    "\"Isg15\"\n",
    "]\n",
    "\n",
    "existing_genes_CD8TG = [g for g in genes_list_CD8TG if g in adata.var_names]\n",
    "missing_genes_CD8TG = [g for g in genes_list_CD8TG if g not in adata.var_names]\n",
    "if missing_genes_CD8TG:\n",
    "    print(\"The following genes were not found:\")\n",
    "    print(missing_genes_CD8TG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c069be2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< subset analysis >>>>>>>\n",
    "\n",
    "for name, subset in subsets:\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs})\")\n",
    "\n",
    "    # ======= pre-analysis of subset =======\n",
    "\n",
    "    subset.raw = subset.copy()\n",
    "\n",
    "    sc.pp.normalize_total(subset, target_sum=1e4)\n",
    "\n",
    "    subset.raw = subset\n",
    "\n",
    "    sc.pp.log1p(subset)\n",
    "\n",
    "    sc.pp.highly_variable_genes(subset, n_top_genes=2000)\n",
    "\n",
    "    sc.pp.scale(subset, max_value=10)\n",
    "\n",
    "    sc.tl.pca(subset)\n",
    "\n",
    "    sc.pp.neighbors(subset)\n",
    "\n",
    "    sc.tl.umap(subset)\n",
    "\n",
    "    sc.tl.leiden(subset, resolution=0.5)\n",
    "\n",
    "\n",
    "    # ======= Cluster =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - Cluster analysis\")\n",
    "\n",
    "    #UMAP-cluster\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"leiden\", \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_cluster_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"leiden\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False, \n",
    "        save=f\"{name}_subset_cluster_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "    groups = deg_result[\"names\"].dtype.names  \n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_cluster_DEG_results.xlsx\") as writer:\n",
    "        for group in groups:\n",
    "            df = pd.DataFrame({\n",
    "                \"gene\": deg_result[\"names\"][group],\n",
    "                \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                \"pvals\": deg_result[\"pvals\"][group],\n",
    "                \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                \"scores\": deg_result[\"scores\"][group],\n",
    "            })\n",
    "            df.to_excel(writer, sheet_name=f\"cluster_{group}\", index=False)\n",
    "    \n",
    "    #UMAP-batch\n",
    "    n_batches = subset.obs[\"batch\"].nunique()\n",
    "    if n_batches >= 2:\n",
    "\n",
    "        subset.obs[\"batch\"] = subset.obs[\"batch\"].astype(\"category\")\n",
    "        subset.obs[\"batch\"] = subset.obs[\"batch\"].cat.reorder_categories(\n",
    "        [\"Control_TG\", \"Control_SP\"], ordered=True\n",
    "        )\n",
    "\n",
    "        custom_palette = [\"#0033ff\", \"#B0B0B0\"] \n",
    "\n",
    "        print(f\"{name}: {n_batches} batches have been detected. Save the UMAP plots color-coded by batch.\")\n",
    "        sc.pl.umap(\n",
    "            subset, \n",
    "            color=\"batch\", \n",
    "            palette=custom_palette, \n",
    "            show=True, \n",
    "            save=f\"{name}_subset_batch_UMAP.pdf\"\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        print(f\"{name}: Since there is only one batch type, batch UMAP has been skipped.\")\n",
    "        \n",
    "\n",
    "    # ======= cdr3 =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - cdr3\")\n",
    "\n",
    "    cdr3_df = pd.read_csv(f\"{cwd}/cdr3_summary_tetTG_-_NaTG.csv\")\n",
    "\n",
    "    cdr3_types = [\n",
    "        (\"CD4TRA\", \"cdr3_TRA\"),\n",
    "        (\"CD4TRB\", \"cdr3_TRB\"),\n",
    "        (\"CD8TRA\", \"cdr3_TRA\"),\n",
    "        (\"CD8TRB\", \"cdr3_TRB\"),\n",
    "    ]\n",
    "\n",
    "    subset.obs[\"cdr3_positive\"] = \"NA\"\n",
    "\n",
    "    for label, subset_col in cdr3_types:\n",
    "        cdr3_list = cdr3_df[f\"{label}_cdr3\"].dropna().unique()\n",
    "        cdr3_set = set(cdr3_list)\n",
    "\n",
    "        condition = (\n",
    "            subset.obs[subset_col].isin(cdr3_set) &\n",
    "            (subset.obs[\"cdr3_positive\"] == \"NA\")  \n",
    "        )\n",
    "\n",
    "        subset.obs.loc[condition, \"cdr3_positive\"] = \"cdr3_true\"\n",
    "\n",
    "    subset.obs[\"cdr3_positive\"] = pd.Categorical(subset.obs[\"cdr3_positive\"], categories=[\"cdr3_true\", \"NA\"])\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"cdr3_positive\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_cdr3_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"cdr3_positive\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_cdr3_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"cdr3_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/mouse_{name}_subset_cdr3_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    # ======= cdr3_TRA or cdr3_TRB =======\n",
    "\n",
    "    for chain, chain_label in [(\"TRA\", \"cdr3_TRA\"), (\"TRB\", \"cdr3_TRB\")]:\n",
    "    \n",
    "        print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - cdr3_{chain} \")\n",
    "\n",
    "        label_col = f\"{chain_label}_positive\"\n",
    "        subset.obs[label_col] = \"NA\"\n",
    "\n",
    "        relevant_cols = [col for col in cdr3_df.columns if col.endswith(f\"{chain}_cdr3\")]\n",
    "        cdr3_combined_set = set()\n",
    "\n",
    "        for col in relevant_cols:\n",
    "            cdr3_combined_set.update(cdr3_df[col].dropna().unique())\n",
    "\n",
    "        condition = subset.obs[chain_label].isin(cdr3_combined_set)\n",
    "        subset.obs.loc[condition, label_col] = f\"{chain_label}_true\"\n",
    "\n",
    "        subset.obs[label_col] = pd.Categorical(subset.obs[label_col])\n",
    "\n",
    "        #UMAP\n",
    "        sc.pl.umap(\n",
    "            subset, \n",
    "            color=label_col, \n",
    "            palette=[\"red\", \"lightgray\"], \n",
    "            show=True, \n",
    "            save=f\"{name}_subset_{chain_label}_UMAP.pdf\"\n",
    "        )\n",
    "\n",
    "        #DEG\n",
    "        sc.tl.rank_genes_groups(\n",
    "            subset, \n",
    "            groupby=label_col, \n",
    "            method=\"wilcoxon\"\n",
    "        )\n",
    "\n",
    "        sc.pl.rank_genes_groups(\n",
    "            subset,\n",
    "            n_genes=25, \n",
    "            sharey=False, \n",
    "            show=False,\n",
    "            save=f\"{name}_subset_{chain_label}_DEG_fig.pdf\"\n",
    "        )\n",
    "\n",
    "        group = f\"{chain_label}_true\"\n",
    "        deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "        df_chain = pd.DataFrame({\n",
    "            \"gene\": deg_result[\"names\"][group],\n",
    "            \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "            \"pvals\": deg_result[\"pvals\"][group],\n",
    "            \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "            \"scores\": deg_result[\"scores\"][group],\n",
    "        })\n",
    "\n",
    "        with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_{chain_label}_DEG_results.xlsx\") as writer:\n",
    "            df_chain.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    # ======= cdr3-cdr3 combo =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - cdr3-cdr3 combo\")\n",
    "\n",
    "    df = pd.read_csv(f\"{cwd}/cdr3_summary_tetTG_-_NaTG_forcdr3combo.csv\")\n",
    "\n",
    "    combo_list = [\n",
    "        f\"{tra}_{trb}\" \n",
    "        for tra in df[\"TRA_cdr3\"].dropna().unique()\n",
    "        for trb in df[\"TRB_cdr3\"].dropna().unique()\n",
    "    ]\n",
    "    \n",
    "    combo_set = set(combo_list)\n",
    "\n",
    "    subset.obs[\"cdr3-cdr3_combo\"] = subset.obs[\"cdr3_TRA_TRB\"].apply(\n",
    "        lambda x: f\"cdr3-cdr3_true\" if x in combo_set else \"NA\"\n",
    "    )\n",
    "\n",
    "    subset.obs[\"cdr3-cdr3_combo\"] = pd.Categorical(subset.obs[\"cdr3-cdr3_combo\"], categories=[\"cdr3-cdr3_true\", \"NA\"])\n",
    "\n",
    "    print(subset.obs[\"cdr3-cdr3_combo\"].cat.categories)\n",
    "\n",
    "    order = subset.obs[\"cdr3-cdr3_combo\"] == \"NA\"\n",
    "    na_cells = subset.obs[order].index\n",
    "    true_cells = subset.obs[~order].index\n",
    "\n",
    "    new_order = na_cells.tolist() + true_cells.tolist()\n",
    "\n",
    "    subset = subset[new_order, :]\n",
    "\n",
    "\n",
    "    true_cells = subset.obs.loc[subset.obs[\"cdr3-cdr3_combo\"] == \"cdr3-cdr3_true\", [\"cdr3_TRA_TRB\", \"v_gene_TRA\", \"j_gene_TRA\", \"cdr3_TRA\", \"v_gene_TRB\", \"j_gene_TRB\", \"cdr3_TRB\"]]\n",
    "\n",
    "    combo_freq = true_cells[\"cdr3_TRA_TRB\"].value_counts().reset_index()\n",
    "    combo_freq.columns = [\"cdr3_combo\", \"frequency\"]\n",
    "\n",
    "    combo_details = true_cells.drop_duplicates(subset=[\"cdr3_TRA_TRB\"]).rename(columns={\"cdr3_TRA_TRB\": \"cdr3_combo\"})\n",
    "\n",
    "    result_df = pd.merge(combo_freq, combo_details, on=\"cdr3_combo\", how=\"left\")\n",
    "\n",
    "    result_df.to_excel(f\"{cwd}/figures/mouse_{name}_subset_cdr3-cdr3_list_frequency.xlsx\", index=False)\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"cdr3-cdr3_combo\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_cdr3-cdr3_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_cdr3-cdr3_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"cdr3-cdr3_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    #dotplot\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset,\n",
    "        var_names=existing_genes_2,\n",
    "        groupby=\"cdr3-cdr3_combo\",\n",
    "        standard_scale='var',\n",
    "        color_map=\"GnBu\",\n",
    "        show=True,\n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_.pdf\")\n",
    "\n",
    "\n",
    "    #dotplot and violin plot\n",
    "    custom_palette = {\"cdr3-cdr3_true\": \"#1f77b4\", \"NA\": \"#B0B0B0\"}\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD4TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD4TGSP.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD4TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD4TGSP.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD4TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD4TG.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD4TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD4TG.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD8TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD8TGSP.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD8TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD8TGSP.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD8TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD8TG.pdf\")\n",
    "    \n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD8TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD8TG.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    #volcano plot\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "    group = \"cdr3-cdr3_true\"\n",
    "    df_deg = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "                \n",
    "    df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "    df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "    def plot_volcano(df, name, cwd, label_genes=True, xlim=None, ylim=None, label_specific_genes=None, label_suffix=None, base_ns_color=\"#B0B0B0\", base_sig_color=\"#d62728\", highlight_color=\"#1f77b4\"):\n",
    "\n",
    "        df = df.copy()\n",
    "\n",
    "        if label_specific_genes is not None:\n",
    "            hi = {g.upper() for g in label_specific_genes}\n",
    "            df[\"is_highlight\"] = df[\"gene\"].str.upper().isin(hi)\n",
    "            genes_to_label = df[df[\"is_highlight\"]]\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"specificgenes\"\n",
    "\n",
    "        elif label_genes:\n",
    "            df[\"is_highlight\"] = df[\"significant\"]\n",
    "            genes_to_label = df[df[\"significant\"]]\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"label\"\n",
    "\n",
    "        else:\n",
    "            df[\"is_highlight\"] = False\n",
    "            genes_to_label = pd.DataFrame()\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"nolabel\"\n",
    "\n",
    "        df[\"_ns\"]  = ~df[\"significant\"]\n",
    "        df[\"_sig\"] = df[\"significant\"] & ~df[\"is_highlight\"]\n",
    "        df[\"_hi\"]  = df[\"is_highlight\"]\n",
    "\n",
    "        def _draw(ax, fix_range=False):\n",
    "            point_size = 24  \n",
    "\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_ns\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_ns\"], \"-log10_pval\"],\n",
    "                c=base_ns_color, s=point_size, alpha=0.7, edgecolors=\"none\", zorder=1\n",
    "            )\n",
    "\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_sig\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_sig\"], \"-log10_pval\"],\n",
    "                c=base_sig_color, s=point_size, alpha=0.9, edgecolors=\"none\", zorder=2\n",
    "            )\n",
    "\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_hi\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_hi\"], \"-log10_pval\"],\n",
    "                c=highlight_color, s=point_size, alpha=1.0, edgecolors=\"none\", zorder=4\n",
    "            )\n",
    "\n",
    "            ax.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "            ax.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "            ax.axvline( 2, linestyle='--', color='black', linewidth=0.8)\n",
    "\n",
    "            if not genes_to_label.empty:\n",
    "\n",
    "                to_annot = genes_to_label.sort_values(\"-log10_pval\", ascending=False)\n",
    "                offsets = [(10, 10)]  \n",
    "\n",
    "                for i, (_, r) in enumerate(to_annot.iterrows()):\n",
    "                    dx, dy = offsets[i % len(offsets)]\n",
    "                    x0, y0 = r[\"logfoldchanges\"], r[\"-log10_pval\"]\n",
    "                    x1, y1 = x0 + dx * 0.05, y0 + dy * 0.05\n",
    "\n",
    "                    ax.text(\n",
    "                        x1, y1, r[\"gene\"],\n",
    "                        fontsize=9, weight=\"bold\", color=\"black\", zorder=6,\n",
    "                        path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]\n",
    "                    )\n",
    "\n",
    "                    ax.plot([x0, x1], [y0, y1], color=\"black\", linewidth=0.8, zorder=5)\n",
    "\n",
    "            ax.set_xlabel(\"log2 Fold Change\")\n",
    "            ax.set_ylabel(\"-log10(p-value)\")\n",
    "            ax.set_title(f\"{name}_subset (cdr3-cdr3_true vs rest)\")\n",
    "\n",
    "\n",
    "            if fix_range:\n",
    "                if xlim is not None: ax.set_xlim(xlim)\n",
    "                if ylim is not None: ax.set_ylim(ylim)\n",
    "                ax.autoscale(enable=False)\n",
    "                ax.margins(x=0, y=0)\n",
    "\n",
    "        plt.figure(figsize=(8, 6))\n",
    "        ax = plt.gca()\n",
    "        _draw(ax, fix_range=False)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_volcano_plot_{label_suffix}.pdf\")\n",
    "        plt.close()\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "        _draw(ax, fix_range=True)\n",
    "        fig.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_volcano_plot_{label_suffix}_fixedrange.pdf\", bbox_inches=\"tight\")\n",
    "        plt.close(fig)\n",
    "\n",
    "\n",
    "\n",
    "    plot_volcano(df_deg, name, cwd, label_genes=True, xlim=[-10, 10])\n",
    "    plot_volcano(df_deg, name, cwd, label_genes=False, xlim=[-10, 10])\n",
    "\n",
    "\n",
    "    specificgenesCD4_TG = [\n",
    "        \"Gbp9\", \"Gbp6\", \"Oas1a\", \"Ifi27l2a\", \"Ifi47\", \"Parp14\", \"Zbp1\", \"Irf7\", \"Isg15\", \"Cxcr3\", \"Icos\", \"Stat1\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD4_TG, \n",
    "        label_suffix = \"specificgenesCD4_TG\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD4_TG_SP = [\n",
    "        \"Itgb1\", \"Icos\", \"Isg15\", \"Bhlhe40\", \"Cxcr3\", \"Ifng\", \"Nr4a1\", \"Gbp6\", \"Oas1a\", \"Isg20\", \"Nkg7\", \"Gbp9\", \"Icam1\", \"Ccl5\", \"Hif1a\", \"Ifi27l2a\", \"Irf7\", \"Zbp1\", \"S1pr1\", \"Il7r\", \"Klf2\", \"Bcl2\", \"Ccr7\", \"Sell\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD4_TG_SP, \n",
    "        label_suffix = \"specificgenesCD4_TG_SP\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD8_DP_DN_TG = [\n",
    "        \"Trbv13-3\", \"Gzmb\", \"Gzma\", \"Itga1\", \"Ccr2\", \"Ly6a\", \"Isg15\", \"Cd8b1\", \"Gzmk\", \"Ccl5\", \"Ctla4\", \"Ifng\", \"Cd8a\", \"S1pr1\", \"Klf2\", \"Bcl2\", \"Tcf7\", \"Ccr7\", \"Sell\", \"Eomes\", \"Tox\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD8_DP_DN_TG, \n",
    "        label_suffix = \"specificgenesCD8_DP_DN_TG\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD8_DP_DN_TG_SP = [\"\"\n",
    "    \"Gzma\", \"Ccl5\", \"S100a6\", \"Gzmb\", \"Trbv13-3\", \"Ifitm1\", \"Itga1\", \"Litaf\", \"Ifng\", \"Gzmk\", \"Ccr2\", \"Bhlhe40\", \"Nr4a2\", \"Ctla4\", \"Ccl4\", \"Ifitm2\", \"Tbx21\", \"Ccr5\", \"Isg15\", \"Cd38\", \"Icos\", \"Cxcr6\", \"Isg20\", \"Nr4a1\", \"Nr4a3\", \"Ifitm3\", \"Nkg7\", \"S1pr1\", \"Klf2\", \"Itga6\", \"Tox\", \"Ccr7\", \"Tcf7\", \"Sell\", \"Ccr9\", \"Tnfsf8\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD8_DP_DN_TG_SP, \n",
    "        label_suffix = \"specificgenesCD8_DP_DN_TG_SP\"\n",
    "    )\n",
    "\n",
    "\n",
    "    # ======= v =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - V\")\n",
    "\n",
    "    V_df = pd.read_csv(f\"{cwd}/selected_V.csv\")\n",
    "\n",
    "    V_types = [\n",
    "        (\"CD4_TRA_V\", \"v_gene_TRA\"),\n",
    "        (\"CD4_TRB_V\", \"v_gene_TRB\"),\n",
    "        (\"CD8_TRA_V\", \"v_gene_TRA\"),\n",
    "        (\"CD8_TRB_V\", \"v_gene_TRB\"),\n",
    "    ]\n",
    "\n",
    "    subset.obs[\"V_positive\"] = \"NA\"\n",
    "\n",
    "    for label, subset_col in V_types:\n",
    "        V_list = V_df[label].dropna().unique()\n",
    "        V_set = set(V_list)\n",
    "\n",
    "        condition = (\n",
    "            subset.obs[subset_col].isin(V_set) &\n",
    "            (subset.obs[\"V_positive\"] == \"NA\")  \n",
    "        )\n",
    "\n",
    "        subset.obs.loc[condition, \"V_positive\"] = \"V_true\"\n",
    "\n",
    "    subset.obs[\"V_positive\"] = pd.Categorical(subset.obs[\"V_positive\"], categories=[\"V_true\", \"NA\"])\n",
    "    print(subset.obs[\"V_positive\"].cat.categories)\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"V_positive\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_V_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"V_positive\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_V_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"V_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_V_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    # ======= v-v combo =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - v-v combo\")\n",
    "\n",
    "    df = pd.read_csv(f\"{cwd}/selected_V_TRA_TRB_forVcombo.csv\")\n",
    "    combo_list = [f\"{tra}_{trb}\" for tra in df[\"TRA_V\"].dropna().unique()\n",
    "                                for trb in df[\"TRB_V\"].dropna().unique()]\n",
    "    combo_set = set(combo_list)\n",
    "\n",
    "    subset.obs[\"V-V_combo\"] = subset.obs[\"V_TRA_TRB\"].apply(\n",
    "        lambda x: f\"V-V_true\" if x in combo_set else \"NA\"\n",
    "    )\n",
    "    subset.obs[\"V-V_combo\"] = pd.Categorical(subset.obs[\"V-V_combo\"], categories=[\"V-V_true\", \"NA\"])\n",
    "    print(subset.obs[\"V-V_combo\"].cat.categories)\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"V-V_combo\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_V-V_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"V-V_combo\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_V-V_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"V-V_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_V-V_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "878eeaba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Define Subsets_2nd for analysis >>>>>>>\n",
    "\n",
    "exclude_clusters = [\"5\"]\n",
    "exclude_cells = subset3.obs[subset3.obs[\"leiden\"].isin(exclude_clusters)].index\n",
    "subset_2nd_1 = subset3[~subset3.obs_names.isin(exclude_cells), :].copy()\n",
    "subset_2nd_2 = subset1[~subset1.obs_names.isin(exclude_cells), :].copy()\n",
    "\n",
    "\n",
    "exclude_clusters = [\"1\", \"5\"]\n",
    "exclude_cells = subset6.obs[subset6.obs[\"leiden\"].isin(exclude_clusters)].index\n",
    "subset_2nd_3 = subset6[~subset6.obs_names.isin(exclude_cells), :].copy()\n",
    "subset_2nd_4 = subset4[~subset4.obs_names.isin(exclude_cells), :].copy()\n",
    "\n",
    "\n",
    "subsets_2nd = [\n",
    "    (\"mouse_CD4_TG+SP_exclude_cluster_5\", subset_2nd_1),\n",
    "    (\"mouse_CD4_TG_exclude_cluster_5\", subset_2nd_2),\n",
    "    (\"mouse_CD8+DP+DN_TG+SP_exclude_cluster_1_5\", subset_2nd_3),\n",
    "    (\"mouse_CD8+DP+DN_TG_exclude_cluster_1_5\", subset_2nd_4),\n",
    "]\n",
    "\n",
    "\n",
    "for name, subset in subsets_2nd:\n",
    "    print(f\"{name}:  {subset.n_obs}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39c0aa9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Subsets_2nd analysis >>>>>>>\n",
    "\n",
    "for name, subset in subsets_2nd:\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset,\n",
    "        color=\"leiden\",\n",
    "        show=True,\n",
    "        save=f\"{name}_subset_cluster_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    cluster_counts = subset.obs[\"leiden\"].value_counts()\n",
    "    valid_clusters = cluster_counts[cluster_counts > 3].index.tolist()\n",
    "        \n",
    "    if len(valid_clusters) >= 2:\n",
    "        subset_filtered = subset[subset.obs[\"leiden\"].isin(valid_clusters), :].copy()\n",
    "        sc.tl.rank_genes_groups(\n",
    "            subset_filtered, \n",
    "            groupby=\"leiden\", \n",
    "            method=\"wilcoxon\"\n",
    "        )\n",
    "        sc.pl.rank_genes_groups(\n",
    "            subset_filtered, \n",
    "            n_genes=25, \n",
    "            sharey=False, \n",
    "            show=False, \n",
    "            save=f\"{name}_subset_cluster_DEG_fig.pdf\"\n",
    "        )\n",
    "\n",
    "        deg_result = subset_filtered.uns[\"rank_genes_groups\"]\n",
    "        groups = deg_result[\"names\"].dtype.names  \n",
    "\n",
    "        with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_cluster_DEG_results.xlsx\") as writer:\n",
    "            for group in groups:\n",
    "                df = pd.DataFrame({\n",
    "                    \"gene\": deg_result[\"names\"][group],\n",
    "                    \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                    \"pvals\": deg_result[\"pvals\"][group],\n",
    "                    \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                    \"scores\": deg_result[\"scores\"][group],\n",
    "                })\n",
    "                df.to_excel(writer, sheet_name=f\"cluster_{group}\", index=False)\n",
    "    else:\n",
    "        print(f\" {name}: DEG analysis was skipped because there were fewer than two valid clusters.\")\n",
    "\n",
    "            \n",
    "\n",
    "    #UMAP-batch\n",
    "    n_batches = subset.obs[\"batch\"].nunique()\n",
    "    if n_batches >= 2:\n",
    "        subset.obs[\"batch\"] = subset.obs[\"batch\"].astype(\"category\")\n",
    "        subset.obs[\"batch\"] = subset.obs[\"batch\"].cat.reorder_categories(\n",
    "        [\"Control_TG\", \"Control_SP\"], ordered=True\n",
    "        )\n",
    "        custom_palette = [\"#0033ff\", \"#B0B0B0\"]  \n",
    "        print(f\"{name}: {n_batches} batches have been detected. Save the UMAP plots color-coded by batch.\")\n",
    "        sc.pl.umap(\n",
    "            subset, \n",
    "            color=\"batch\", \n",
    "            palette=custom_palette, \n",
    "            show=True, \n",
    "            save=f\"{name}_subset_batch_UMAP.pdf\"\n",
    "        )\n",
    "    else:\n",
    "        print(f\"{name}: Since there is only one batch type, batch UMAP has been skipped.\")\n",
    "\n",
    "\n",
    "\n",
    "    # ======= cdr3-cdr3 combo =======\n",
    "\n",
    "    order = subset.obs[\"cdr3-cdr3_combo\"] == \"NA\"\n",
    "    na_cells = subset.obs[order].index\n",
    "    true_cells = subset.obs[~order].index\n",
    "\n",
    "    new_order = na_cells.tolist() + true_cells.tolist()\n",
    "\n",
    "    subset = subset[new_order, :]\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"cdr3-cdr3_combo\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_cdr3-cdr3_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    cluster_counts = subset.obs[\"leiden\"].value_counts()\n",
    "    valid_clusters = cluster_counts[cluster_counts > 3].index.tolist()\n",
    "\n",
    "    if len(valid_clusters) >= 2:\n",
    "\n",
    "        subset_filtered = subset[subset.obs[\"leiden\"].isin(valid_clusters), :].copy()\n",
    "        sc.tl.rank_genes_groups(\n",
    "            subset_filtered, \n",
    "            groupby=\"cdr3-cdr3_combo\", \n",
    "            method=\"wilcoxon\"\n",
    "        )\n",
    "        sc.pl.rank_genes_groups(\n",
    "            subset_filtered, \n",
    "            n_genes=25, \n",
    "            sharey=False, \n",
    "            show=False , \n",
    "            save=f\"{name}_subset_cdr3-cdr3_DEG_fig.pdf\"\n",
    "        )\n",
    "\n",
    "        deg_result = subset_filtered.uns[\"rank_genes_groups\"]\n",
    "\n",
    "        group = \"cdr3-cdr3_true\"\n",
    "        df = pd.DataFrame({\n",
    "            \"gene\": deg_result[\"names\"][group],\n",
    "            \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "            \"pvals\": deg_result[\"pvals\"][group],\n",
    "            \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "            \"scores\": deg_result[\"scores\"][group],\n",
    "        })\n",
    "\n",
    "        with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_DEG_results.xlsx\") as writer:\n",
    "            df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "            \n",
    "    else:\n",
    "        print(f\"{name}: DEG analysis was skipped because there were fewer than two valid clusters.\")\n",
    "\n",
    "\n",
    "\n",
    "    #dotplot\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset_filtered,\n",
    "        var_names=existing_genes_2,\n",
    "        groupby=\"cdr3-cdr3_combo\",\n",
    "        standard_scale='var',\n",
    "        color_map=\"GnBu\",\n",
    "        show=True,\n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_.pdf\")\n",
    "\n",
    "    \n",
    "    #dotplot and violin plot\n",
    "    custom_palette = {\"cdr3-cdr3_true\": \"#1f77b4\", \"NA\": \"#B0B0B0\"}\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD4TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD4TGSP.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD4TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD4TGSP.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD4TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD4TG.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD4TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD4TG.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD8TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD8TGSP.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD8TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD8TGSP.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD8TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD8TG.pdf\")\n",
    "    \n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD8TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD8TG.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    #volcano plot\n",
    "    deg_result = subset_filtered.uns[\"rank_genes_groups\"]\n",
    "    group = \"cdr3-cdr3_true\"\n",
    "    df_deg = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "\n",
    "    df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "    df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "    def plot_volcano(df, name, cwd, label_genes=True, xlim=None, ylim=None, label_specific_genes=None, label_suffix=None, base_ns_color=\"#B0B0B0\", base_sig_color=\"#d62728\", highlight_color=\"#1f77b4\"):\n",
    "\n",
    "        df = df.copy()\n",
    "\n",
    "        if label_specific_genes is not None:\n",
    "            hi = {g.upper() for g in label_specific_genes}\n",
    "            df[\"is_highlight\"] = df[\"gene\"].str.upper().isin(hi)\n",
    "            genes_to_label = df[df[\"is_highlight\"]]\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"specificgenes\"\n",
    "\n",
    "        elif label_genes:\n",
    "            df[\"is_highlight\"] = df[\"significant\"]\n",
    "            genes_to_label = df[df[\"significant\"]]\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"label\"\n",
    "\n",
    "        else:\n",
    "            df[\"is_highlight\"] = False\n",
    "            genes_to_label = pd.DataFrame()\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"nolabel\"\n",
    "\n",
    "        df[\"_ns\"]  = ~df[\"significant\"]\n",
    "        df[\"_sig\"] = df[\"significant\"] & ~df[\"is_highlight\"]\n",
    "        df[\"_hi\"]  = df[\"is_highlight\"]\n",
    "\n",
    "        def _draw(ax, fix_range=False):\n",
    "            point_size = 24 \n",
    "\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_ns\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_ns\"], \"-log10_pval\"],\n",
    "                c=base_ns_color, s=point_size, alpha=0.7, edgecolors=\"none\", zorder=1\n",
    "            )\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_sig\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_sig\"], \"-log10_pval\"],\n",
    "                c=base_sig_color, s=point_size, alpha=0.9, edgecolors=\"none\", zorder=2\n",
    "            )\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_hi\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_hi\"], \"-log10_pval\"],\n",
    "                c=highlight_color, s=point_size, alpha=1.0, edgecolors=\"none\", zorder=4\n",
    "            )\n",
    "\n",
    "            ax.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "            ax.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "            ax.axvline( 2, linestyle='--', color='black', linewidth=0.8)\n",
    "\n",
    "            if not genes_to_label.empty:\n",
    "                to_annot = genes_to_label.sort_values(\"-log10_pval\", ascending=False)\n",
    "                offsets = [(10, 10)]  \n",
    "                for i, (_, r) in enumerate(to_annot.iterrows()):\n",
    "                    dx, dy = offsets[i % len(offsets)]\n",
    "                    x0, y0 = r[\"logfoldchanges\"], r[\"-log10_pval\"]\n",
    "                    x1, y1 = x0 + dx * 0.05, y0 + dy * 0.05\n",
    "\n",
    "                    ax.text(\n",
    "                        x1, y1, r[\"gene\"],\n",
    "                        fontsize=9, weight=\"bold\", color=\"black\", zorder=6,\n",
    "                        path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]\n",
    "                    )\n",
    "                    ax.plot([x0, x1], [y0, y1], color=\"black\", linewidth=0.8, zorder=5)\n",
    "\n",
    "            ax.set_xlabel(\"log2 Fold Change\")\n",
    "            ax.set_ylabel(\"-log10(p-value)\")\n",
    "            ax.set_title(f\"{name}_subset (cdr3-cdr3_true vs rest)\")\n",
    "\n",
    "\n",
    "            if fix_range:\n",
    "                if xlim is not None: ax.set_xlim(xlim)\n",
    "                if ylim is not None: ax.set_ylim(ylim)\n",
    "                ax.autoscale(enable=False)\n",
    "                ax.margins(x=0, y=0)\n",
    "\n",
    "        plt.figure(figsize=(8, 6))\n",
    "        ax = plt.gca()\n",
    "        _draw(ax, fix_range=False)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_volcano_plot_{label_suffix}.pdf\")\n",
    "        plt.close()\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "        _draw(ax, fix_range=True)\n",
    "        fig.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_volcano_plot_{label_suffix}_fixedrange.pdf\", bbox_inches=\"tight\")\n",
    "        plt.close(fig)\n",
    "\n",
    "\n",
    "\n",
    "    plot_volcano(df_deg, name, cwd, label_genes=True, xlim=[-10, 10])\n",
    "    plot_volcano(df_deg, name, cwd, label_genes=False, xlim=[-10, 10])\n",
    "\n",
    "\n",
    "\n",
    "    specificgenesCD4_TG = [\n",
    "        \"Gbp9\", \"Gbp6\", \"Oas1a\", \"Ifi27l2a\", \"Ifi47\", \"Parp14\", \"Zbp1\", \"Irf7\", \"Isg15\", \"Cxcr3\", \"Icos\", \"Stat1\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD4_TG, \n",
    "        label_suffix = \"specificgenesCD4_TG\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD4_TG_SP = [\n",
    "        \"Itgb1\", \"Icos\", \"Isg15\", \"Bhlhe40\", \"Cxcr3\", \"Ifng\", \"Nr4a1\", \"Gbp6\", \"Oas1a\", \"Isg20\", \"Nkg7\", \"Gbp9\", \"Icam1\", \"Ccl5\", \"Hif1a\", \"Ifi27l2a\", \"Irf7\", \"Zbp1\", \"S1pr1\", \"Il7r\", \"Klf2\", \"Bcl2\", \"Ccr7\", \"Sell\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD4_TG_SP, \n",
    "        label_suffix = \"specificgenesCD4_TG_SP\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD8_DP_DN_TG = [\n",
    "        \"Trbv13-3\", \"Gzmb\", \"Gzma\", \"Itga1\", \"Ccr2\", \"Ly6a\", \"Isg15\", \"Cd8b1\", \"Gzmk\", \"Ccl5\", \"Ctla4\", \"Ifng\", \"Cd8a\", \"S1pr1\", \"Klf2\", \"Bcl2\", \"Tcf7\", \"Ccr7\", \"Sell\", \"Eomes\", \"Tox\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD8_DP_DN_TG, \n",
    "        label_suffix = \"specificgenesCD8_DP_DN_TG\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD8_DP_DN_TG_SP = [\"\"\n",
    "    \"Gzma\", \"Ccl5\", \"S100a6\", \"Gzmb\", \"Trbv13-3\", \"Ifitm1\", \"Itga1\", \"Litaf\", \"Ifng\", \"Gzmk\", \"Ccr2\", \"Bhlhe40\", \"Nr4a2\", \"Ctla4\", \"Ccl4\", \"Ifitm2\", \"Tbx21\", \"Ccr5\", \"Isg15\", \"Cd38\", \"Icos\", \"Cxcr6\", \"Isg20\", \"Nr4a1\", \"Nr4a3\", \"Ifitm3\", \"Nkg7\", \"S1pr1\", \"Klf2\", \"Itga6\", \"Tox\", \"Ccr7\", \"Tcf7\", \"Sell\", \"Ccr9\", \"Tnfsf8\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD8_DP_DN_TG_SP, \n",
    "        label_suffix = \"specificgenesCD8_DP_DN_TG_SP\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "242d46bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# ======= \"CD4_TG\", subset1 =======\n",
    "\n",
    "target_combos = [\n",
    "    \"CAVSAGYGSSGNKLIF_CASSYQGATGQLYF\",\n",
    "]\n",
    "\n",
    "cdr3_true_cells = subset1[subset1.obs[\"cdr3-cdr3_combo\"] == \"cdr3-cdr3_true\", :].copy()\n",
    "\n",
    "cdr3_true_cells.obs[\"combo_specific\"] = np.where(\n",
    "    cdr3_true_cells.obs[\"cdr3_TRA_TRB\"].isin(target_combos),\n",
    "    \"combo_match\",\n",
    "    \"combo_other\"\n",
    ")\n",
    "cdr3_true_cells.obs[\"combo_specific\"] = pd.Categorical(cdr3_true_cells.obs[\"combo_specific\"], categories=[\"combo_match\", \"combo_other\"])\n",
    "\n",
    "sc.tl.rank_genes_groups(\n",
    "    cdr3_true_cells, \n",
    "    groupby=\"combo_specific\", \n",
    "    method=\"wilcoxon\"\n",
    ")\n",
    "\n",
    "deg_result = cdr3_true_cells.uns[\"rank_genes_groups\"]\n",
    "group = \"combo_match\"\n",
    "\n",
    "df_deg = pd.DataFrame({\n",
    "    \"gene\": deg_result[\"names\"][group],\n",
    "    \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "    \"pvals\": deg_result[\"pvals\"][group],\n",
    "    \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "    \"scores\": deg_result[\"scores\"][group],\n",
    "})\n",
    "df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "df_deg.to_excel(f\"{cwd}/figures/mouse_CD4_TG_subset_target-cdr3-combo_vs_others_DEG_result.xlsx\", index=False)\n",
    "\n",
    "df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "def plot_volcano(df, name, cwd, label_genes=True, xlim=[-10, 10], ylim=None):\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    plt.scatter(\n",
    "        df[\"logfoldchanges\"], df[\"-log10_pval\"],\n",
    "        c=df[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "        alpha=0.7, edgecolors=\"none\"\n",
    "    )\n",
    "    if label_genes:\n",
    "        for _, row in df[df[\"significant\"]].iterrows():\n",
    "            plt.text(\n",
    "                row[\"logfoldchanges\"],\n",
    "                row[\"-log10_pval\"],\n",
    "                row[\"gene\"],\n",
    "                fontsize=8,\n",
    "                ha='center',\n",
    "                va='bottom'\n",
    "            )\n",
    "    plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)  \n",
    "    plt.axvline(-2, linestyle='--', color='black', linewidth=0.8) \n",
    "    plt.axvline(2, linestyle='--', color='black', linewidth=0.8)   \n",
    "    plt.xlabel(\"log2 Fold Change\")\n",
    "    plt.ylabel(\"-log10(p-value)\")\n",
    "    plt.title(f\"{name}_subset (cdr3-cdr3_true vs rest)\")\n",
    "    plt.tight_layout()\n",
    "\n",
    "    label_suffix = \"label\" if label_genes else \"nolabel\"\n",
    "    plt.savefig(f\"{cwd}/figures/mouse_CD4_TG_subset_target-cdr3-combo_vs_others_volcano_{label_suffix}.pdf\")\n",
    "    plt.close()\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)                    \n",
    "    plt.scatter(\n",
    "        df[\"logfoldchanges\"], df[\"-log10_pval\"],\n",
    "        c=df[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "        alpha=0.7, edgecolors=\"none\"\n",
    "    )\n",
    "    if label_genes:\n",
    "        for _, row in df[df[\"significant\"]].iterrows():\n",
    "            plt.text(\n",
    "                row[\"logfoldchanges\"],\n",
    "                row[\"-log10_pval\"],\n",
    "                row[\"gene\"],\n",
    "                fontsize=8,\n",
    "                ha='center',\n",
    "                va='bottom'\n",
    "            )\n",
    "\n",
    "    plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "    plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "    plt.axvline(2, linestyle='--', color='black', linewidth=0.8)\n",
    "    plt.xlabel(\"log2 Fold Change\")\n",
    "    plt.ylabel(\"-log10(p-value)\")\n",
    "    plt.title(f\"{name}_subset (cdr3-cdr3_true vs rest)\")\n",
    "\n",
    "    if xlim is not None:\n",
    "        plt.xlim(xlim)\n",
    "    if ylim is not None:\n",
    "        plt.ylim(ylim)\n",
    "    ax.autoscale(enable=False)  \n",
    "    ax.margins(x=0, y=0) \n",
    "\n",
    "    plt.savefig(f\"{cwd}/figures/mouse_CD4_TG_subset_target-cdr3-combo_vs_others_volcano_{label_suffix}_fixedrange.pdf\", bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "\n",
    "plot_volcano(df_deg, name, cwd, label_genes=True)\n",
    "plot_volcano(df_deg, name, cwd, label_genes=False)\n",
    "\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "plt.figure(figsize=(9, 6))\n",
    "plt.scatter(\n",
    "    df_deg[\"logfoldchanges\"], df_deg[\"-log10_pval\"],\n",
    "    c=df_deg[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "    alpha=0.7, edgecolors=\"none\"\n",
    ")\n",
    "plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(2, linestyle='--', color='black', linewidth=0.8)\n",
    "\n",
    "texts = []\n",
    "for _, row in df_deg[df_deg[\"significant\"]].iterrows():\n",
    "    texts.append(plt.text(row[\"logfoldchanges\"], row[\"-log10_pval\"], row[\"gene\"], fontsize=8))\n",
    "\n",
    "plt.xlabel(\"log2 Fold Change\")\n",
    "plt.ylabel(\"-log10(p-value)\")\n",
    "plt.title(f\"Volcano Plot: CD4_TG  target-cdr3-combo vs others\")\n",
    "plt.xlim(-10, 10)\n",
    "ax.autoscale(enable=False)  \n",
    "ax.margins(x=0, y=0) \n",
    "\n",
    "\n",
    "combo_text = \"\\n\".join(target_combos)\n",
    "plt.text(\n",
    "    0.95, 0.05,\n",
    "    f\"Target combos:\\n{combo_text}\",\n",
    "    ha='right', va='bottom',\n",
    "    fontsize=8,\n",
    "    transform=plt.gca().transAxes,\n",
    "    bbox=dict(boxstyle=\"round\", facecolor=\"white\", alpha=0.6)\n",
    ")\n",
    "\n",
    "plt.tight_layout(rect=[0, 0.1, 1, 1])\n",
    "plt.savefig(f\"{cwd}/figures/mouse_CD4_TG_subset_target-cdr3-combo_vs_others_volcano_label_2.pdf\")\n",
    "plt.close()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "plt.figure(figsize=(9, 6))\n",
    "plt.scatter(\n",
    "    df_deg[\"logfoldchanges\"], df_deg[\"-log10_pval\"],\n",
    "    c=df_deg[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "    alpha=0.7, edgecolors=\"none\"\n",
    ")\n",
    "plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(2, linestyle='--', color='black', linewidth=0.8)\n",
    "plt.xlabel(\"log2 Fold Change\")\n",
    "plt.ylabel(\"-log10(p-value)\")\n",
    "plt.title(f\"Volcano Plot: CD4_TG  target-cdr3-combo vs others\")\n",
    "plt.xlim(-10, 10)\n",
    "ax.autoscale(enable=False)  \n",
    "ax.margins(x=0, y=0) \n",
    "\n",
    "plt.text(\n",
    "    0.95, 0.05,\n",
    "    f\"Target combos:\\n{combo_text}\",\n",
    "    ha='right', va='bottom',\n",
    "    fontsize=8,\n",
    "    transform=plt.gca().transAxes,\n",
    "    bbox=dict(boxstyle=\"round\", facecolor=\"white\", alpha=0.6)\n",
    ")\n",
    "\n",
    "plt.tight_layout(rect=[0, 0.1, 1, 1])\n",
    "plt.savefig(f\"{cwd}/figures/mouse_CD4_TG_subset_target-cdr3-combo_vs_others_volcano_nolabel_2.pdf\")\n",
    "plt.close()\n",
    "\n",
    "\n",
    "\n",
    "sc.pl.dotplot(\n",
    "    cdr3_true_cells,\n",
    "    var_names=existing_genes_2,\n",
    "    groupby=\"combo_specific\",\n",
    "    standard_scale='var',\n",
    "    color_map=\"GnBu\",\n",
    "    show=True,          \n",
    "    save = \"mouse_CD4_TG_subset_target-cdr3-combo_vs_others_dotplot.pdf\"     \n",
    ")\n",
    "\n",
    "\n",
    "# ======= \"CD8+DP+DN_TG\", subset4 =======\n",
    "\n",
    "\n",
    "target_combos = [\n",
    "    \"CAASSNYNVLYF_CASRGQGAETLYF\",\n",
    "    \"CAASGSNNRIFF_CASRGQGQNTLYF\",\n",
    "    \"CAASTNYNVLYF_CASRGQGISDYTF\",\n",
    "    \"CAASAHYSNNRLTL_CASSLRGSGNTLYF\",\n",
    "    \"CAAGGQGTGSKLSF_CASSYWGSQNTLYF\",\n",
    "    \"CALGSNYNVLYF_CASRGQGNTEVFF\",\n",
    "]\n",
    "\n",
    "cdr3_true_cells = subset4[subset4.obs[\"cdr3-cdr3_combo\"] == \"cdr3-cdr3_true\", :].copy()\n",
    "\n",
    "cdr3_true_cells.obs[\"combo_specific\"] = np.where(\n",
    "    cdr3_true_cells.obs[\"cdr3_TRA_TRB\"].isin(target_combos),\n",
    "    \"combo_match\",\n",
    "    \"combo_other\"\n",
    ")\n",
    "cdr3_true_cells.obs[\"combo_specific\"] = pd.Categorical(cdr3_true_cells.obs[\"combo_specific\"], categories=[\"combo_match\", \"combo_other\"])\n",
    "\n",
    "sc.tl.rank_genes_groups(\n",
    "    cdr3_true_cells, \n",
    "    groupby=\"combo_specific\", \n",
    "    method=\"wilcoxon\"\n",
    ")\n",
    "\n",
    "deg_result = cdr3_true_cells.uns[\"rank_genes_groups\"]\n",
    "group = \"combo_match\"\n",
    "\n",
    "df_deg = pd.DataFrame({\n",
    "    \"gene\": deg_result[\"names\"][group],\n",
    "    \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "    \"pvals\": deg_result[\"pvals\"][group],\n",
    "    \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "    \"scores\": deg_result[\"scores\"][group],\n",
    "})\n",
    "df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "df_deg.to_excel(f\"{cwd}/figures/mouse_CD8+DP+DN_TG_subset_target-cdr3-combo_vs_others_DEG_result.xlsx\", index=False)\n",
    "\n",
    "df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "def plot_volcano(df, name, cwd, label_genes=True, xlim=[-10, 10], ylim=None):\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    plt.scatter(\n",
    "        df[\"logfoldchanges\"], df[\"-log10_pval\"],\n",
    "        c=df[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "        alpha=0.7, edgecolors=\"none\"\n",
    "    )\n",
    "    if label_genes:\n",
    "        for _, row in df[df[\"significant\"]].iterrows():\n",
    "            plt.text(\n",
    "                row[\"logfoldchanges\"],\n",
    "                row[\"-log10_pval\"],\n",
    "                row[\"gene\"],\n",
    "                fontsize=8,\n",
    "                ha='center',\n",
    "                va='bottom'\n",
    "            )\n",
    "    plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)  \n",
    "    plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)  \n",
    "    plt.axvline(2, linestyle='--', color='black', linewidth=0.8)   \n",
    "    plt.xlabel(\"log2 Fold Change\")\n",
    "    plt.ylabel(\"-log10(p-value)\")\n",
    "    plt.title(f\"{name}_subset (cdr3-cdr3_true vs rest)\")\n",
    "    plt.tight_layout()\n",
    "\n",
    "    label_suffix = \"label\" if label_genes else \"nolabel\"\n",
    "    plt.savefig(f\"{cwd}/figures/mouse_CD8+DP+DN_TG_subset_target-cdr3-combo_vs_others_volcano_{label_suffix}.pdf\")\n",
    "    plt.close()\n",
    "\n",
    "    fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)                    \n",
    "    plt.scatter(\n",
    "        df[\"logfoldchanges\"], df[\"-log10_pval\"],\n",
    "        c=df[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "        alpha=0.7, edgecolors=\"none\"\n",
    "    )\n",
    "    if label_genes:\n",
    "        for _, row in df[df[\"significant\"]].iterrows():\n",
    "            plt.text(\n",
    "                row[\"logfoldchanges\"],\n",
    "                row[\"-log10_pval\"],\n",
    "                row[\"gene\"],\n",
    "                fontsize=8,\n",
    "                ha='center',\n",
    "                va='bottom'\n",
    "            )\n",
    "\n",
    "    plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "    plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "    plt.axvline(2, linestyle='--', color='black', linewidth=0.8)\n",
    "    plt.xlabel(\"log2 Fold Change\")\n",
    "    plt.ylabel(\"-log10(p-value)\")\n",
    "    plt.title(f\"{name}_subset (cdr3-cdr3_true vs rest)\")\n",
    "\n",
    "    if xlim is not None:\n",
    "        plt.xlim(xlim)\n",
    "    if ylim is not None:\n",
    "        plt.ylim(ylim)\n",
    "    ax.autoscale(enable=False)  \n",
    "    ax.margins(x=0, y=0) \n",
    "\n",
    "    plt.savefig(f\"{cwd}/figures/mouse_CD8+DP+DN_TG_subset_target-cdr3-combo_vs_others_volcano_{label_suffix}_fixedrange.pdf\", bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "\n",
    "plot_volcano(df_deg, name, cwd, label_genes=True)\n",
    "plot_volcano(df_deg, name, cwd, label_genes=False)\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "\n",
    "plt.figure(figsize=(9, 6))\n",
    "plt.scatter(\n",
    "    df_deg[\"logfoldchanges\"], df_deg[\"-log10_pval\"],\n",
    "    c=df_deg[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "    alpha=0.7, edgecolors=\"none\"\n",
    ")\n",
    "plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(2, linestyle='--', color='black', linewidth=0.8)\n",
    "\n",
    "texts = []\n",
    "for _, row in df_deg[df_deg[\"significant\"]].iterrows():\n",
    "    texts.append(plt.text(row[\"logfoldchanges\"], row[\"-log10_pval\"], row[\"gene\"], fontsize=8))\n",
    "\n",
    "plt.xlabel(\"log2 Fold Change\")\n",
    "plt.ylabel(\"-log10(p-value)\")\n",
    "plt.title(f\"Volcano Plot: CD8+DP+DN_TG  target-cdr3-combo vs others\")\n",
    "plt.xlim(-10, 10)\n",
    "ax.autoscale(enable=False)  \n",
    "ax.margins(x=0, y=0) \n",
    "\n",
    "combo_text = \"\\n\".join(target_combos)\n",
    "plt.text(\n",
    "    0.95, 0.05,\n",
    "    f\"Target combos:\\n{combo_text}\",\n",
    "    ha='right', va='bottom',\n",
    "    fontsize=8,\n",
    "    transform=plt.gca().transAxes,\n",
    "    bbox=dict(boxstyle=\"round\", facecolor=\"white\", alpha=0.6)\n",
    ")\n",
    "\n",
    "plt.tight_layout(rect=[0, 0.1, 1, 1])\n",
    "plt.savefig(f\"{cwd}/figures/mouse_CD8+DP+DN_TG_subset_target-cdr3-combo_vs_others_volcano_label_2.pdf\")\n",
    "plt.close()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "\n",
    "plt.figure(figsize=(9, 6))\n",
    "plt.scatter(\n",
    "    df_deg[\"logfoldchanges\"], df_deg[\"-log10_pval\"],\n",
    "    c=df_deg[\"significant\"].map({True: \"red\", False: \"gray\"}),\n",
    "    alpha=0.7, edgecolors=\"none\"\n",
    ")\n",
    "plt.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "plt.axvline(2, linestyle='--', color='black', linewidth=0.8)\n",
    "plt.xlabel(\"log2 Fold Change\")\n",
    "plt.ylabel(\"-log10(p-value)\")\n",
    "plt.title(f\"Volcano Plot: CD8+DP+DN_TG  target-cdr3-combo vs others\")\n",
    "plt.xlim(-10, 10)\n",
    "ax.autoscale(enable=False)  \n",
    "ax.margins(x=0, y=0) \n",
    "\n",
    "plt.text(\n",
    "    0.95, 0.05,\n",
    "    f\"Target combos:\\n{combo_text}\",\n",
    "    ha='right', va='bottom',\n",
    "    fontsize=8,\n",
    "    transform=plt.gca().transAxes,\n",
    "    bbox=dict(boxstyle=\"round\", facecolor=\"white\", alpha=0.6)\n",
    ")\n",
    "\n",
    "plt.tight_layout(rect=[0, 0.1, 1, 1])\n",
    "plt.savefig(f\"{cwd}/figures/mouse_CD8+DP+DN_TG_subset_target-cdr3-combo_vs_others_volcano_nolabel_2.pdf\")\n",
    "plt.close()\n",
    "\n",
    "\n",
    "sc.pl.dotplot(\n",
    "    cdr3_true_cells,\n",
    "    var_names=existing_genes_2,\n",
    "    groupby=\"combo_specific\",\n",
    "    standard_scale='var',\n",
    "    color_map=\"GnBu\",\n",
    "    show=True,  \n",
    "    save = \"mouse_CD8+DP+DN_TG_subset_target-cdr3-combo_vs_others_dotplot.pdf\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d6f4c8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ======================= gene expression dotplot (subsets) =======================\n",
    "\n",
    "\n",
    "\n",
    "subsets_of_a_cell = []\n",
    "for name, subset in subsets:\n",
    "    subset_cp = subset.copy()\n",
    "    subset_cp.obs[\"subset_name\"] = name\n",
    "    subset_cp.obs_names = [f\"{name}_{i}\" for i in subset_cp.obs_names]  \n",
    "    subsets_of_a_cell.append(subset_cp)\n",
    "\n",
    "adata_subsets_of_a_cell = anndata.concat(subsets_of_a_cell, join=\"outer\")\n",
    "\n",
    "dp = sc.pl.dotplot(\n",
    "    adata_subsets_of_a_cell,\n",
    "    var_names=existing_genes,\n",
    "    groupby=\"subset_name\",\n",
    "    standard_scale='var',\n",
    "    color_map=\"GnBu\",\n",
    "    show=True,\n",
    "    return_fig=True\n",
    ")\n",
    "\n",
    "dp.savefig(f\"{cwd}/figures/mouse_dotplot_by_subset.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1218069",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ======================= gene expression dotplot (each subset) =======================\n",
    "\n",
    "\n",
    "for i, gene_list in enumerate([existing_genes, existing_genes_2]):\n",
    "\n",
    "    for name, subset in subsets:\n",
    "        dp = sc.pl.dotplot(\n",
    "            subset,\n",
    "            var_names=gene_list,\n",
    "            groupby=\"leiden\",\n",
    "            standard_scale='var',\n",
    "            color_map=\"GnBu\",\n",
    "            show=True,\n",
    "            return_fig=True\n",
    "        )\n",
    "\n",
    "        dp.savefig(f\"{cwd}/figures/{name}_subset_cluster_dotplot_{i}.pdf\")\n",
    "\n",
    "\n",
    "    for name, subset in subsets_2nd:\n",
    "        dp = sc.pl.dotplot(\n",
    "            subset,\n",
    "            var_names=gene_list,\n",
    "            groupby=\"leiden\",\n",
    "            standard_scale='var',\n",
    "            color_map=\"GnBu\",\n",
    "            show=True,\n",
    "            return_fig=True\n",
    "        )\n",
    "\n",
    "        dp.savefig(f\"{cwd}/figures/{name}_subset_cluster_dotplot_{i}.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57696a83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Define Subsets_3rd for analysis >>>>>>>\n",
    "\n",
    "subset_3rd_1 = subset3[\n",
    "    (subset3.obs[\"leiden\"].isin([\"2\"]))\n",
    "].copy()   \n",
    "\n",
    "\n",
    "subset_3rd_2 = subset6[\n",
    "    (subset6.obs[\"leiden\"].isin([\"2\"]))\n",
    "].copy()\n",
    "\n",
    "\n",
    "subsets_3rd = [\n",
    "    (\"mouse_CD4_TG_SP_Cluster2\", subset_3rd_1),\n",
    "    (\"mouse_CD8+DP+DN_TG_SP_Cluster2\", subset_3rd_2),\n",
    "]\n",
    "\n",
    "for name, subset in subsets_3rd:\n",
    "    print(f\"{name}:  {subset.n_obs}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57688c6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# <<<<<<< Subsets_3rd analysis >>>>>>>\n",
    "\n",
    "for name, subset in subsets_3rd:\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs})\")\n",
    "\n",
    "    # ======= pre-analysis of subset =======\n",
    "\n",
    "    subset = subset.raw.to_adata()\n",
    "\n",
    "    sc.pp.normalize_total(subset, target_sum=1e4)\n",
    "\n",
    "    subset.raw = subset\n",
    "\n",
    "    sc.pp.log1p(subset)\n",
    "\n",
    "    sc.pp.highly_variable_genes(subset, n_top_genes=2000)\n",
    "\n",
    "    sc.pp.scale(subset, max_value=10)\n",
    "\n",
    "    sc.tl.pca(subset)\n",
    "\n",
    "    sc.pp.neighbors(subset)\n",
    "\n",
    "    sc.tl.umap(subset)\n",
    "\n",
    "    sc.tl.leiden(subset, resolution=0.5)\n",
    "\n",
    "\n",
    "    # ======= Cluster =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - Cluster analysis\")\n",
    "\n",
    "    #UMAP-cluster\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"leiden\", \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_cluster_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"leiden\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False, \n",
    "        save=f\"{name}_subset_cluster_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "    groups = deg_result[\"names\"].dtype.names  \n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_cluster_DEG_results.xlsx\") as writer:\n",
    "        for group in groups:\n",
    "            df = pd.DataFrame({\n",
    "                \"gene\": deg_result[\"names\"][group],\n",
    "                \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "                \"pvals\": deg_result[\"pvals\"][group],\n",
    "                \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "                \"scores\": deg_result[\"scores\"][group],\n",
    "            })\n",
    "            df.to_excel(writer, sheet_name=f\"cluster_{group}\", index=False)\n",
    "    \n",
    "    #UMAP-batch\n",
    "    n_batches = subset.obs[\"batch\"].nunique()\n",
    "    if n_batches >= 2:\n",
    "\n",
    "        subset.obs[\"batch\"] = subset.obs[\"batch\"].astype(\"category\")\n",
    "        subset.obs[\"batch\"] = subset.obs[\"batch\"].cat.reorder_categories(\n",
    "        [\"Control_TG\", \"Control_SP\"], ordered=True\n",
    "        )\n",
    "\n",
    "        custom_palette = [\"#0033ff\", \"#B0B0B0\"] \n",
    "\n",
    "        print(f\"{name}: {n_batches} batches have been detected. Save the UMAP plots color-coded by batch.\")\n",
    "        sc.pl.umap(\n",
    "            subset, \n",
    "            color=\"batch\", \n",
    "            palette=custom_palette, \n",
    "            show=True, \n",
    "            save=f\"{name}_subset_batch_UMAP.pdf\"\n",
    "        )\n",
    "\n",
    "    else:\n",
    "        print(f\"{name}: Since there is only one batch type, batch UMAP has been skipped.\")\n",
    "        \n",
    "\n",
    "    # ======= cdr3 =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - cdr3\")\n",
    "\n",
    "    cdr3_df = pd.read_csv(f\"{cwd}/cdr3_summary_tetTG_-_NaTG.csv\")\n",
    "\n",
    "    cdr3_types = [\n",
    "        (\"CD4TRA\", \"cdr3_TRA\"),\n",
    "        (\"CD4TRB\", \"cdr3_TRB\"),\n",
    "        (\"CD8TRA\", \"cdr3_TRA\"),\n",
    "        (\"CD8TRB\", \"cdr3_TRB\"),\n",
    "    ]\n",
    "\n",
    "    subset.obs[\"cdr3_positive\"] = \"NA\"\n",
    "\n",
    "    for label, subset_col in cdr3_types:\n",
    "        cdr3_list = cdr3_df[f\"{label}_cdr3\"].dropna().unique()\n",
    "        cdr3_set = set(cdr3_list)\n",
    "\n",
    "        condition = (\n",
    "            subset.obs[subset_col].isin(cdr3_set) &\n",
    "            (subset.obs[\"cdr3_positive\"] == \"NA\")  \n",
    "        )\n",
    "\n",
    "        subset.obs.loc[condition, \"cdr3_positive\"] = \"cdr3_true\"\n",
    "\n",
    "    subset.obs[\"cdr3_positive\"] = pd.Categorical(subset.obs[\"cdr3_positive\"], categories=[\"cdr3_true\", \"NA\"])\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"cdr3_positive\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_cdr3_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"cdr3_positive\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_cdr3_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"cdr3_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/mouse_{name}_subset_cdr3_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    # ======= cdr3_TRA or cdr3_TRB =======\n",
    "\n",
    "    for chain, chain_label in [(\"TRA\", \"cdr3_TRA\"), (\"TRB\", \"cdr3_TRB\")]:\n",
    "    \n",
    "        print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - cdr3_{chain} \")\n",
    "\n",
    "        label_col = f\"{chain_label}_positive\"\n",
    "        subset.obs[label_col] = \"NA\"\n",
    "\n",
    "        relevant_cols = [col for col in cdr3_df.columns if col.endswith(f\"{chain}_cdr3\")]\n",
    "        cdr3_combined_set = set()\n",
    "\n",
    "        for col in relevant_cols:\n",
    "            cdr3_combined_set.update(cdr3_df[col].dropna().unique())\n",
    "\n",
    "        condition = subset.obs[chain_label].isin(cdr3_combined_set)\n",
    "        subset.obs.loc[condition, label_col] = f\"{chain_label}_true\"\n",
    "\n",
    "        subset.obs[label_col] = pd.Categorical(subset.obs[label_col])\n",
    "\n",
    "        #UMAP\n",
    "        sc.pl.umap(\n",
    "            subset, \n",
    "            color=label_col, \n",
    "            palette=[\"red\", \"lightgray\"], \n",
    "            show=True, \n",
    "            save=f\"{name}_subset_{chain_label}_UMAP.pdf\"\n",
    "        )\n",
    "\n",
    "        #DEG\n",
    "        sc.tl.rank_genes_groups(\n",
    "            subset, \n",
    "            groupby=label_col, \n",
    "            method=\"wilcoxon\"\n",
    "        )\n",
    "\n",
    "        sc.pl.rank_genes_groups(\n",
    "            subset,\n",
    "            n_genes=25, \n",
    "            sharey=False, \n",
    "            show=False,\n",
    "            save=f\"{name}_subset_{chain_label}_DEG_fig.pdf\"\n",
    "        )\n",
    "\n",
    "        group = f\"{chain_label}_true\"\n",
    "        deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "        df_chain = pd.DataFrame({\n",
    "            \"gene\": deg_result[\"names\"][group],\n",
    "            \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "            \"pvals\": deg_result[\"pvals\"][group],\n",
    "            \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "            \"scores\": deg_result[\"scores\"][group],\n",
    "        })\n",
    "\n",
    "        with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_{chain_label}_DEG_results.xlsx\") as writer:\n",
    "            df_chain.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    # ======= cdr3-cdr3 combo =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - cdr3-cdr3 combo\")\n",
    "\n",
    "    df = pd.read_csv(f\"{cwd}/cdr3_summary_tetTG_-_NaTG_forcdr3combo.csv\")\n",
    "\n",
    "    combo_list = [\n",
    "        f\"{tra}_{trb}\" \n",
    "        for tra in df[\"TRA_cdr3\"].dropna().unique()\n",
    "        for trb in df[\"TRB_cdr3\"].dropna().unique()\n",
    "    ]\n",
    "    \n",
    "    combo_set = set(combo_list)\n",
    "\n",
    "    subset.obs[\"cdr3-cdr3_combo\"] = subset.obs[\"cdr3_TRA_TRB\"].apply(\n",
    "        lambda x: f\"cdr3-cdr3_true\" if x in combo_set else \"NA\"\n",
    "    )\n",
    "\n",
    "    subset.obs[\"cdr3-cdr3_combo\"] = pd.Categorical(subset.obs[\"cdr3-cdr3_combo\"], categories=[\"cdr3-cdr3_true\", \"NA\"])\n",
    "\n",
    "    print(subset.obs[\"cdr3-cdr3_combo\"].cat.categories)\n",
    "\n",
    "    order = subset.obs[\"cdr3-cdr3_combo\"] == \"NA\"\n",
    "    na_cells = subset.obs[order].index\n",
    "    true_cells = subset.obs[~order].index\n",
    "\n",
    "    new_order = na_cells.tolist() + true_cells.tolist()\n",
    "\n",
    "    subset = subset[new_order, :]\n",
    "\n",
    "\n",
    "    true_cells = subset.obs.loc[subset.obs[\"cdr3-cdr3_combo\"] == \"cdr3-cdr3_true\", [\"cdr3_TRA_TRB\", \"v_gene_TRA\", \"j_gene_TRA\", \"cdr3_TRA\", \"v_gene_TRB\", \"j_gene_TRB\", \"cdr3_TRB\"]]\n",
    "\n",
    "    combo_freq = true_cells[\"cdr3_TRA_TRB\"].value_counts().reset_index()\n",
    "    combo_freq.columns = [\"cdr3_combo\", \"frequency\"]\n",
    "\n",
    "    combo_details = true_cells.drop_duplicates(subset=[\"cdr3_TRA_TRB\"]).rename(columns={\"cdr3_TRA_TRB\": \"cdr3_combo\"})\n",
    "\n",
    "    result_df = pd.merge(combo_freq, combo_details, on=\"cdr3_combo\", how=\"left\")\n",
    "\n",
    "    result_df.to_excel(f\"{cwd}/figures/mouse_{name}_subset_cdr3-cdr3_list_frequency.xlsx\", index=False)\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"cdr3-cdr3_combo\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_cdr3-cdr3_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_cdr3-cdr3_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"cdr3-cdr3_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    #dotplot\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset,\n",
    "        var_names=existing_genes_2,\n",
    "        groupby=\"cdr3-cdr3_combo\",\n",
    "        standard_scale='var',\n",
    "        color_map=\"GnBu\",\n",
    "        show=True,\n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_.pdf\")\n",
    "\n",
    "\n",
    "    #dotplot and violin plot\n",
    "    custom_palette = {\"cdr3-cdr3_true\": \"#1f77b4\", \"NA\": \"#B0B0B0\"}\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD4TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD4TGSP.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD4TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD4TGSP.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD4TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD4TG.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD4TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD4TG.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD8TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD8TGSP.pdf\")\n",
    "\n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD8TGSP, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD8TGSP.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    dp = sc.pl.dotplot(\n",
    "        subset, \n",
    "        var_names=existing_genes_CD8TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        standard_scale='var', \n",
    "        color_map=\"GnBu\", \n",
    "        show=True, \n",
    "        return_fig=True\n",
    "    )\n",
    "    dp.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_dotplot_CD8TG.pdf\")\n",
    "    \n",
    "    sc.pl.violin(\n",
    "        subset, \n",
    "        keys=existing_genes_CD8TG, \n",
    "        groupby=\"cdr3-cdr3_combo\", \n",
    "        palette=custom_palette, \n",
    "        stripplot=True, \n",
    "        jitter=0.25, \n",
    "        size=2, \n",
    "        rotation=90, \n",
    "        show=True, \n",
    "        save=f\"_{name}_subset_cdr3-cdr3_violinplot_CD8TG.pdf\"\n",
    "    )\n",
    "\n",
    "\n",
    "    #volcano plot\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "    group = \"cdr3-cdr3_true\"\n",
    "    df_deg = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "                \n",
    "    df_deg[\"-log10_pval\"] = -np.log10(df_deg[\"pvals\"])\n",
    "    df_deg[\"significant\"] = (df_deg[\"pvals\"] < 0.05) & (abs(df_deg[\"logfoldchanges\"]) > 2)\n",
    "\n",
    "    def plot_volcano(df, name, cwd, label_genes=True, xlim=None, ylim=None, label_specific_genes=None, label_suffix=None, base_ns_color=\"#B0B0B0\", base_sig_color=\"#d62728\", highlight_color=\"#1f77b4\"):\n",
    "\n",
    "        df = df.copy()\n",
    "\n",
    "        if label_specific_genes is not None:\n",
    "            hi = {g.upper() for g in label_specific_genes}\n",
    "            df[\"is_highlight\"] = df[\"gene\"].str.upper().isin(hi)\n",
    "            genes_to_label = df[df[\"is_highlight\"]]\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"specificgenes\"\n",
    "\n",
    "        elif label_genes:\n",
    "            df[\"is_highlight\"] = df[\"significant\"]\n",
    "            genes_to_label = df[df[\"significant\"]]\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"label\"\n",
    "\n",
    "        else:\n",
    "            df[\"is_highlight\"] = False\n",
    "            genes_to_label = pd.DataFrame()\n",
    "            if label_suffix is None:\n",
    "                label_suffix = \"nolabel\"\n",
    "\n",
    "        df[\"_ns\"]  = ~df[\"significant\"]\n",
    "        df[\"_sig\"] = df[\"significant\"] & ~df[\"is_highlight\"]\n",
    "        df[\"_hi\"]  = df[\"is_highlight\"]\n",
    "\n",
    "        def _draw(ax, fix_range=False):\n",
    "            point_size = 24  \n",
    "\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_ns\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_ns\"], \"-log10_pval\"],\n",
    "                c=base_ns_color, s=point_size, alpha=0.7, edgecolors=\"none\", zorder=1\n",
    "            )\n",
    "\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_sig\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_sig\"], \"-log10_pval\"],\n",
    "                c=base_sig_color, s=point_size, alpha=0.9, edgecolors=\"none\", zorder=2\n",
    "            )\n",
    "\n",
    "            ax.scatter(\n",
    "                df.loc[df[\"_hi\"], \"logfoldchanges\"],\n",
    "                df.loc[df[\"_hi\"], \"-log10_pval\"],\n",
    "                c=highlight_color, s=point_size, alpha=1.0, edgecolors=\"none\", zorder=4\n",
    "            )\n",
    "\n",
    "            ax.axhline(-np.log10(0.05), linestyle='--', color='black', linewidth=0.8)\n",
    "            ax.axvline(-2, linestyle='--', color='black', linewidth=0.8)\n",
    "            ax.axvline( 2, linestyle='--', color='black', linewidth=0.8)\n",
    "\n",
    "            if not genes_to_label.empty:\n",
    "\n",
    "                to_annot = genes_to_label.sort_values(\"-log10_pval\", ascending=False)\n",
    "                offsets = [(10, 10)]  \n",
    "\n",
    "                for i, (_, r) in enumerate(to_annot.iterrows()):\n",
    "                    dx, dy = offsets[i % len(offsets)]\n",
    "                    x0, y0 = r[\"logfoldchanges\"], r[\"-log10_pval\"]\n",
    "                    x1, y1 = x0 + dx * 0.05, y0 + dy * 0.05\n",
    "\n",
    "                    ax.text(\n",
    "                        x1, y1, r[\"gene\"],\n",
    "                        fontsize=9, weight=\"bold\", color=\"black\", zorder=6,\n",
    "                        path_effects=[pe.withStroke(linewidth=3, foreground=\"white\")]\n",
    "                    )\n",
    "\n",
    "                    ax.plot([x0, x1], [y0, y1], color=\"black\", linewidth=0.8, zorder=5)\n",
    "\n",
    "            ax.set_xlabel(\"log2 Fold Change\")\n",
    "            ax.set_ylabel(\"-log10(p-value)\")\n",
    "            ax.set_title(f\"{name}_subset (cdr3-cdr3_true vs rest)\")\n",
    "\n",
    "\n",
    "            if fix_range:\n",
    "                if xlim is not None: ax.set_xlim(xlim)\n",
    "                if ylim is not None: ax.set_ylim(ylim)\n",
    "                ax.autoscale(enable=False)\n",
    "                ax.margins(x=0, y=0)\n",
    "\n",
    "        plt.figure(figsize=(8, 6))\n",
    "        ax = plt.gca()\n",
    "        _draw(ax, fix_range=False)\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_volcano_plot_{label_suffix}.pdf\")\n",
    "        plt.close()\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(8, 6), constrained_layout=False)\n",
    "        _draw(ax, fix_range=True)\n",
    "        fig.savefig(f\"{cwd}/figures/{name}_subset_cdr3-cdr3_volcano_plot_{label_suffix}_fixedrange.pdf\", bbox_inches=\"tight\")\n",
    "        plt.close(fig)\n",
    "\n",
    "\n",
    "\n",
    "    plot_volcano(df_deg, name, cwd, label_genes=True, xlim=[-10, 10])\n",
    "    plot_volcano(df_deg, name, cwd, label_genes=False, xlim=[-10, 10])\n",
    "\n",
    "\n",
    "    specificgenesCD4_TG = [\n",
    "        \"Gbp9\", \"Gbp6\", \"Oas1a\", \"Ifi27l2a\", \"Ifi47\", \"Parp14\", \"Zbp1\", \"Irf7\", \"Isg15\", \"Cxcr3\", \"Icos\", \"Stat1\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD4_TG, \n",
    "        label_suffix = \"specificgenesCD4_TG\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD4_TG_SP = [\n",
    "        \"Itgb1\", \"Icos\", \"Isg15\", \"Bhlhe40\", \"Cxcr3\", \"Ifng\", \"Nr4a1\", \"Gbp6\", \"Oas1a\", \"Isg20\", \"Nkg7\", \"Gbp9\", \"Icam1\", \"Ccl5\", \"Hif1a\", \"Ifi27l2a\", \"Irf7\", \"Zbp1\", \"S1pr1\", \"Il7r\", \"Klf2\", \"Bcl2\", \"Ccr7\", \"Sell\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD4_TG_SP, \n",
    "        label_suffix = \"specificgenesCD4_TG_SP\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD8_DP_DN_TG = [\n",
    "        \"Trbv13-3\", \"Gzmb\", \"Gzma\", \"Itga1\", \"Ccr2\", \"Ly6a\", \"Isg15\", \"Cd8b1\", \"Gzmk\", \"Ccl5\", \"Ctla4\", \"Ifng\", \"Cd8a\", \"S1pr1\", \"Klf2\", \"Bcl2\", \"Tcf7\", \"Ccr7\", \"Sell\", \"Eomes\", \"Tox\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD8_DP_DN_TG, \n",
    "        label_suffix = \"specificgenesCD8_DP_DN_TG\"\n",
    "    )\n",
    "\n",
    "\n",
    "    specificgenesCD8_DP_DN_TG_SP = [\"\"\n",
    "    \"Gzma\", \"Ccl5\", \"S100a6\", \"Gzmb\", \"Trbv13-3\", \"Ifitm1\", \"Itga1\", \"Litaf\", \"Ifng\", \"Gzmk\", \"Ccr2\", \"Bhlhe40\", \"Nr4a2\", \"Ctla4\", \"Ccl4\", \"Ifitm2\", \"Tbx21\", \"Ccr5\", \"Isg15\", \"Cd38\", \"Icos\", \"Cxcr6\", \"Isg20\", \"Nr4a1\", \"Nr4a3\", \"Ifitm3\", \"Nkg7\", \"S1pr1\", \"Klf2\", \"Itga6\", \"Tox\", \"Ccr7\", \"Tcf7\", \"Sell\", \"Ccr9\", \"Tnfsf8\"\n",
    "    ]\n",
    "    plot_volcano(\n",
    "        df_deg, \n",
    "        name, \n",
    "        cwd, \n",
    "        label_genes=False, \n",
    "        xlim=[-10, 10], \n",
    "        label_specific_genes=specificgenesCD8_DP_DN_TG_SP, \n",
    "        label_suffix = \"specificgenesCD8_DP_DN_TG_SP\"\n",
    "    )\n",
    "\n",
    "\n",
    "    # ======= v =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - V\")\n",
    "\n",
    "    V_df = pd.read_csv(f\"{cwd}/selected_V.csv\")\n",
    "\n",
    "    V_types = [\n",
    "        (\"CD4_TRA_V\", \"v_gene_TRA\"),\n",
    "        (\"CD4_TRB_V\", \"v_gene_TRB\"),\n",
    "        (\"CD8_TRA_V\", \"v_gene_TRA\"),\n",
    "        (\"CD8_TRB_V\", \"v_gene_TRB\"),\n",
    "    ]\n",
    "\n",
    "    subset.obs[\"V_positive\"] = \"NA\"\n",
    "\n",
    "    for label, subset_col in V_types:\n",
    "        V_list = V_df[label].dropna().unique()\n",
    "        V_set = set(V_list)\n",
    "\n",
    "        condition = (\n",
    "            subset.obs[subset_col].isin(V_set) &\n",
    "            (subset.obs[\"V_positive\"] == \"NA\")  \n",
    "        )\n",
    "\n",
    "        subset.obs.loc[condition, \"V_positive\"] = \"V_true\"\n",
    "\n",
    "    subset.obs[\"V_positive\"] = pd.Categorical(subset.obs[\"V_positive\"], categories=[\"V_true\", \"NA\"])\n",
    "    print(subset.obs[\"V_positive\"].cat.categories)\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"V_positive\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_V_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"V_positive\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_V_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"V_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_V_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "\n",
    "\n",
    "    # ======= v-v combo =======\n",
    "\n",
    "    print(f\"Processing: {name}(Num of cells: {subset.n_obs}) - v-v combo\")\n",
    "\n",
    "    df = pd.read_csv(f\"{cwd}/selected_V_TRA_TRB_forVcombo.csv\")\n",
    "    combo_list = [f\"{tra}_{trb}\" for tra in df[\"TRA_V\"].dropna().unique()\n",
    "                                for trb in df[\"TRB_V\"].dropna().unique()]\n",
    "    combo_set = set(combo_list)\n",
    "\n",
    "    subset.obs[\"V-V_combo\"] = subset.obs[\"V_TRA_TRB\"].apply(\n",
    "        lambda x: f\"V-V_true\" if x in combo_set else \"NA\"\n",
    "    )\n",
    "    subset.obs[\"V-V_combo\"] = pd.Categorical(subset.obs[\"V-V_combo\"], categories=[\"V-V_true\", \"NA\"])\n",
    "    print(subset.obs[\"V-V_combo\"].cat.categories)\n",
    "\n",
    "    #UMAP\n",
    "    sc.pl.umap(\n",
    "        subset, \n",
    "        color=\"V-V_combo\", \n",
    "        palette=[\"red\", \"lightgray\"], \n",
    "        show=True, \n",
    "        save=f\"{name}_subset_V-V_UMAP.pdf\"\n",
    "    )\n",
    "\n",
    "    #DEG\n",
    "    sc.tl.rank_genes_groups(\n",
    "        subset, \n",
    "        groupby=\"V-V_combo\", \n",
    "        method=\"wilcoxon\"\n",
    "    )\n",
    "\n",
    "    sc.pl.rank_genes_groups(\n",
    "        subset, \n",
    "        n_genes=25, \n",
    "        sharey=False, \n",
    "        show=False , \n",
    "        save=f\"{name}_subset_V-V_DEG_fig.pdf\"\n",
    "    )\n",
    "\n",
    "    deg_result = subset.uns[\"rank_genes_groups\"]\n",
    "\n",
    "    group = \"V-V_true\"\n",
    "    df = pd.DataFrame({\n",
    "        \"gene\": deg_result[\"names\"][group],\n",
    "        \"logfoldchanges\": deg_result[\"logfoldchanges\"][group],\n",
    "        \"pvals\": deg_result[\"pvals\"][group],\n",
    "        \"pvals_adj\": deg_result[\"pvals_adj\"][group],\n",
    "        \"scores\": deg_result[\"scores\"][group],\n",
    "    })\n",
    "\n",
    "    with pd.ExcelWriter(f\"{cwd}/figures/{name}_subset_V-V_DEG_results.xlsx\") as writer:\n",
    "        df.to_excel(writer, sheet_name=f\"{group}_vs_rest\", index=False)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02b1d1c",
   "metadata": {},
   "source": [
    "# Rename in Tarminal\n",
    "\n",
    "for f in *mouse*; do\n",
    "  new=$(echo \"$f\" | sed 's/^.*mouse/mouse/')\n",
    "  mv \"$f\" \"$new\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94bf41a8",
   "metadata": {},
   "source": [
    "# Rename in PowerShell\n",
    "\n",
    "Get-ChildItem -File *mouse* | ForEach-Object {\n",
    "    $newName = $_.Name -replace '.*mouse','mouse'\n",
    "    if ($_.Name -ne $newName) {\n",
    "        Rename-Item -LiteralPath $_.FullName -NewName $newName\n",
    "        Write-Output \"$($_.Name)  $newName\"\n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Aomine_scRNA-seq_0718",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
